---
title: "Introduction"
execute:
  warning: false
  message: false
  eval: false
editor: 
  markdown: 
    wrap: 72
---

This repository contains the teaching materials of **Single Cell Boot
Camp - Analyzing scRNA-seq with R**. The boot camp aims to provide an
overview of scRNAseq analysis workflow. It begins with a fast-paced
session on the analysis workflow and proceeds with hands-on practice
using bioinformatics packages such as Seurat.

## Prerequisites to attend the Single Cell Boot Camp

Each participant are expected to have:

The following statistical concepts:

-   median: A measure of central tendency that represents the middle
    value in a dataset when arranged in ascending order.

-   log2 fold change: A measure used in gene expression analysis to
    quantify the difference in expression levels between two conditions.

-   p-value: A statistical measure that helps determine the significance
    of a result in the context of a hypothesis test.

-   adjusted p-value: A corrected p-value that accounts for multiple
    comparisons to control for false positive.

-   variance: A measure of how much the values in a dataset deviate from
    the mean.

-   principal component analysis (PCA): A technique used to reduce the
    dimensionality of high-dimensional data while capturing most of its
    variance.

Additionally, participants should have a basic understanding of R and
the tidyverse package. Familiarity with the following functions will be
useful:

-   `%>%` (pipe): Allows for chaining multiple operations together.

-   `filter`: Selects specific rows from a dataset on given conditions.

-   `mutate`: Adds new variables or modifies existing ones in a dataset.

-   `select`: Selects specific columns from a dataset.

-   `left_join`: Merges two datasets based on a common key, keeping only
    the matching records from the left dataset.

For self-learning of R programming and the tidyverse package, it is
recommended study [Chapter 1 - 4](https://moderndive.com/index.html)
from MordernDive.

## Expected outcome

By the end of the workshop, participants will have a solid understanding
of the analysis workflow involved in single cell analysis. This includes
data pre-processing, quality control, normalization, dimentionality
reduction, clustering, and cell type annotation.

## Reference

The building of materials references and relies on many excellent and
well-written public available materials. Some of the main references for
the course include:

1.  Seurat <https://satijalab.org/seurat/>
2.  Single Cell Best Practices:
    <https://www.sc-best-practices.org/preamble.html>
3.  Orchestrating Single-Cell Analysis with Bioconductor:
    <https://bioconductor.org/books/release/OSCA/>
4.  Mary Piper, Meeta Mistry, Jihe Liu, William Gammerdinger, & Radhika
    Khetani. (2022, January 6). hbctraining/scRNA-seq_online: scRNA-seq
    Lessons from HCBC (first release). Zenodo.
    <https://doi.org/10.5281/zenodo.5826256>

## Reporting bugs

To report any bugs, please create an issue in the following GitHub
repository:
<https://github.com/material-share/material-share-single-cell-boot-camp/issues>.

## Disclaimer

Recommendations given by representatives of BD in the context of the
creation and setup of customer-specific applications and assays, the
composition of antibody panels and suggestions for data analysis are
provided in a diligent manner. However, the Customer acknowledges that
(i) such recommendations should not be treated as a substitute for the
Customer's own examination, (ii) BD does not make any promises nor
guarantees that recommendations are accurate and complete and will meet
the Customer's present or future needs or will produce positive or
specific results and (iii) Customer remains solely responsible for
validating and deciding on any such recommendations.

BD therefore accepts no liability whatsoever in connection with any such
recommendation and/or any results generated on the basis of such
recommendations and, to the fullest extent permitted by law, expressly
disclaims any and all warranties and any liability in this respect. In
particular, the Customer assumes sole responsibility for the validation,
use, selection, and suitability of the recommendations for its needs and
objectives.

## License

MIT License

Copyright (c) 2023 material-share

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
