---
title: "Chapter 1 Data Import"
execute:
  warning: false
  message: false
---

```{r}
#| message: false
# libraries for this chapter
library(tidyverse)
library(Seurat)
library(data.table)
```

# 1. Loading single cell data

```{r}
expMat <- Read10X("data/raw_data/WTA-AB-SMK-v2-0_DBEC_MolsPerCell_MEX/",
                  gene.column = 2,
                  cell.column = 1,
                  unique.features = TRUE)
```

Click on the `expMat` on the right.

![](pics/1-1-expMat.png)

A window will pop up and you should be able to see the structure of `expMat`.

![](pics/1-2-expMat.png)

We will first focus on **Gene Expression**.

```{r}
expMat$`Gene Expression`[1:5, 1:5]
```

::: callout-important
# @QUESTION1

What are the dots "." in the matrix?
:::

# 2. Create a Seurat object to store Gene Expression

```{r}
rna <- Seurat::CreateSeuratObject(counts = expMat$`Gene Expression`, 
                                min.cells = 1, 
                                min.features = 100, 
                                assay = "RNA")
```

::: callout-important
# @QUESTION2

How many genes and cells are stored in the Seurat object **`rna`**?

::: callout-tip
## Tip

Click on the Seurat object **`rna`** in the top right window. What is **28127 x 4948**?

![](pics/Screenshot%202023-09-17%20at%2008.15.13.png)
:::
:::

# 3. Add sample tag information into `meta.data`

```{r}
# load sample tag info into R
smk <- fread(file = "data/raw_data/WTA-AB-SMK-v2-0_Sample_Tag_Calls.csv", 
                     sep = ",", 
                     header=TRUE) %>% 
  data.frame(row.names = 1)
```

```{r}
# add sample tag info into seurat object
rna <- AddMetaData(object = rna, metadata = smk)
```

::: callout-important
# @QUESTION3

How do we know the sample tag information is added into the Seurat object `rna`?

::: callout-tip
## Tip

Click on **`rna`** in the top right window, then click on `meta.data`. What do you see?

![](pics/Screenshot%202023-09-17%20at%2008.15.13%202.png)
:::
:::
