---
title: "Chapter 1 Data Import"
execute:
  warning: false
  message: false
format:
  html:
    embed-resources: true
---

```{r}
# libraries for this chapter
library(tidyverse)
library(Seurat)
library(data.table)
```

# Demo data

The demo data is generated from freshly isolated human PBMCs. The PBMCs were grown for 24 hours in the presence of CD3/CD28 stimulation ("treated") and compared to PBMCs grown for 24 hours in regular media ("naive"). Cells were then labeled with a 20 plex AbSeq panel consisting of lineage markers and T-cell activation markers, as well as unique Sample Tags from the BD Rhapsody^™^ Human Single Cell Multiplexing Kit (resting PBMC: ST1; treated PBMC: ST2). The cells were mixed at a 1:1 ratio and processed using the BD Rhapsody^™^ Single-Cell Analysis System. Sequencing libraries for RNA and AbSeq antibodies were generated using the BD Rhapsody WTA Amplification kit.

::: callout-note
Please review the **setup** chapter to install the R packages and download the demo dataset.
:::

# 1. Data import

Rhapsody single cell platform outputs a rds file containing the expression matrix as a `Seurat` object. The expression matrix stores the UMI counts of a feature (gene or AbSeq).

Let's import the rds file in RStudio.

```{r}
seuratObj <- readRDS("processed_BD-Demo-WTA-AbSeq-SMK_Seurat.rds")
```

Click on the `seuratObj` on the right. A window will pop up and you should be able to see the structure of `seuratObj`.

`seuratObj` contains two assays: **RNA** and **ADT**. **RNA** stores **Gene Expression**; **ADT** stores **AbSeq**.

`meta.data` stores information related to each cell. We will add more information into the `meta.data` later. For now, it has the below columns:

-   `orig.ident` Sample identity or sample name. By default, it is "SeuratProject".

-   `nCount_RNA` UMI counts per gene per cell.

-   `nFeature_RNA` Genes counts per cell.

-   `Cell_Type_Experiment` Cell annotation based on 100 marker genes of immune cell types.

-   `Sample_Tag` The cell is from which sample tag.

-   `Sample_Name` The cell is from which sample.

-   `Putative_Cell_Algorithm` Which algorithm is used to call the cell. Default is "Basic".

-   `nCount_ADT` UMI counts per AB per cell.

-   `nFeature_ADT` AB counts per cell.

```{r}
#| echo: false

# save the seurat object
save(seuratObj, file = "data/chapter1.RData")
```
