<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">

<head>

<meta charset="utf-8" />
<meta name="generator" content="quarto-1.3.450" />

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />


<title>Chapter 6 Marker Identification</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<!-- htmldependencies:E3FAD763 -->
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css" />
</head>

<body>

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn"
      data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" 
      aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation"
      onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <h1 class="quarto-secondary-nav-title"></h1>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" 
      aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation"
      onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="/">
      Single Cell Boot Camp: Analyzing Single Cell RNA-seq with R
      </a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Setting up</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/precourse_preparation_R.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Precourse Preparation - R</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/precourse_preparation_theory.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Precourse Preparation - Theory</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/01_data_import.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 1 Data Import</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/02_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 2 Quality Control</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/03_normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 3 Normalization and PCA</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/04_integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 4 Integration</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/05_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 5 Clustering</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/06_marker_identification.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Chapter 6 Marker Identification</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/08_enrichment_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 8 Enrichment Analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/09_combining.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 9 Combine scRNAseq and AbSeq</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/supplementary_targeted_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Supplementary: Data import and quality control for targeted assays</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/sessioninfo.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">SessionInfo</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" ></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <div id="quarto-toc-target"></div>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Chapter 6 Marker Identification</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>
<nav id="TOC" role="doc-toc">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#load-rdata" id="toc-load-rdata">Load RData</a></li>
  <li><a href="#find-marker-genes" id="toc-find-marker-genes">1. Find marker genes</a>
  <ul>
  <li><a href="#findallmarkers" id="toc-findallmarkers">1.1 <code>FindAllMarkers</code></a></li>
  <li><a href="#findconservedmarkers" id="toc-findconservedmarkers">1.2 <code>FindConservedMarkers</code></a></li>
  <li><a href="#findmarkers" id="toc-findmarkers">1.3 <code>FindMarkers</code></a></li>
  </ul></li>
  <li><a href="#automatic-annotation" id="toc-automatic-annotation">2. Automatic annotation</a>
  <ul>
  <li><a href="#reference-based" id="toc-reference-based">2.1 Reference-based</a></li>
  <li><a href="#marker-based" id="toc-marker-based">2.1 Marker-based</a></li>
  </ul></li>
  <li><a href="#save" id="toc-save">3. Save</a></li>
  </ul>
</nav>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-1_59966eaa4a76fdeb24ee8feda826e068">
<div class="sourceCode" id="cb1"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># libraries for this chapter</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(HGNChelper)</span></code></pre></div>
</div>
<section id="load-rdata" class="level1">
<h1>Load RData</h1>
<p>Load the RData which contains all files from the previous chapter.</p>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-2_38dd3572e2382bc4cda1f5fa67a0e553">
<div class="sourceCode" id="cb2"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;data/intermediate_data/chapter5.RData&quot;</span>)</span></code></pre></div>
</div>
</section>
<section id="find-marker-genes" class="level1">
<h1>1. Find marker genes</h1>
<section id="findallmarkers" class="level2">
<h2>1.1 <code>FindAllMarkers</code></h2>
<p>This type of analysis is typically for evaluating a single sample group/condition. With the <code>FindAllMarkers</code> function, we are comparing each cluster to all other clusters to identify potential marker genes. The cells in each cluster are treated as replicates, and essentially a differential expression analysis is performed with some statistical test. By default, the test is Wilcoxon Rank Sum test.</p>
<p>The <code>FindAllMarkers</code> function has three arguments to determine if a gene is a marker gene.</p>
<ul>
<li><p><code>logfc.threshold</code> Limit testing to genes that show, on average, at least an X-fold difference (log-scale) between the two groups of cells. The default is 0.25. Increasing <code>logfc.threshold</code> speeds up the function but can miss weaker signals.</p></li>
<li><p><code>min.pct</code> Only test genes that are detected in a minimum fraction of cells in either of the two populations. It is meant to speed up the function by not testing infrequently expressed genes. The default is 0.1.</p></li>
<li><p><code>only.pos</code> Only return positive markers (FALSE by default).</p></li>
</ul>
<p>One may change the arguments depending on how stringent to be.</p>
<p>We look at the marker genes in the clusters in sample “treated”.</p>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-3_3d309c48297f8adfc7f2262ec5e23be5">
<div class="sourceCode" id="cb3"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>treated.markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(<span class="fu">subset</span>(integrate.filtered.rna, <span class="at">subset =</span> Sample_Name <span class="sc">==</span> <span class="st">&quot;treated&quot;</span>), </span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">assay =</span> <span class="st">&quot;RNA&quot;</span>,</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">only.pos =</span> <span class="cn">TRUE</span>, </span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">min.pct =</span> <span class="fl">0.25</span>, </span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span></code></pre></div>
</div>
<p>Let’s look at the top 2 significant marker genes in each cluster.</p>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-4_d6a5768d4bc4ed9c8792446e4bae4831">
<div class="sourceCode" id="cb4"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>treated.markers <span class="sc">%&gt;%</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">slice_max</span>(<span class="at">n =</span> <span class="dv">2</span>, <span class="at">order_by =</span> avg_log2FC)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 26 × 7
# Groups:   cluster [13]
       p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene      
       &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt;     
 1 9.89e-158       2.63 0.744 0.212 1.91e-153 0       MAL       
 2 3.62e-104       2.10 0.734 0.379 7.01e-100 0       PCED1B-AS1
 3 0               5.06 0.961 0.055 0         1       GNLY      
 4 3.65e-259       4.13 0.892 0.108 7.07e-255 1       CCL5      
 5 5.52e- 12       1.07 0.75  0.515 1.07e-  7 2       HSPD1     
 6 6.86e- 17       1.00 0.957 0.787 1.33e- 12 2       NPM1      
 7 5.53e- 82       2.41 0.594 0.119 1.07e- 77 3       RGS1      
 8 5.81e- 25       1.82 0.304 0.09  1.12e- 20 3       FOXP3     
 9 1.24e-122       2.25 0.916 0.276 2.39e-118 4       LTA       
10 1.65e-159       1.97 0.941 0.225 3.20e-155 4       MIR155HG  
# ℹ 16 more rows</code></pre>
</div>
</div>
<p>We get a table with the information on marker genes.</p>
<p><code>p_val</code> P-value before multiple testing.</p>
<p><code>avg_log2FC</code> Log2 fold change on the average gene expression between one cluster and the rest of cluster combined.</p>
<p><code>pct.1</code> Percentage of cells expressing the gene in one group. Here is one cluster.</p>
<p><code>pct.2</code> Percentage of cells expressing the gene in another group. Here is the rest of cluster combined.</p>
<p><code>p_val_adj</code> p-value with multiple testing correction.</p>
</section>
<section id="findconservedmarkers" class="level2">
<h2>1.2 <code>FindConservedMarkers</code></h2>
<p>We have two samples, naïve and treated. We can find conserved markers between the conditions with the function, <code>FindConservedMarkers</code>.</p>
<p>This function internally separates out cells by sample group/condition, then performs differential gene expression testing for a single specified cluster against all other clusters.</p>
<p>Gene-level p-values are computed for each condition and then combined across groups using meta-analysis methods from the <code>MetaDE</code> R package.</p>
<p>Many of the arguments in <code>FindConservedMarkers</code> are shared with <code>FindAllMarkers</code>. Here, we list some additional arguments when using <code>FindConservedMarkers</code>.</p>
<ul>
<li><p><code>ident.1</code> This function only evaluates one cluster at a time; here, you would specify the cluster of interest.</p></li>
<li><p><code>grouping.var</code> The variable (column name) in your metadata specifies the separation of cells into groups.</p></li>
</ul>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-5_9342c7f74876e97e712d02fcf809915f">
<div class="sourceCode" id="cb6"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>fcm <span class="ot">&lt;-</span> <span class="fu">FindConservedMarkers</span>(integrate.filtered.rna,</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">assay =</span> <span class="st">&quot;RNA&quot;</span>,</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">ident.1 =</span> <span class="st">&quot;1&quot;</span>,</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">grouping.var =</span> <span class="st">&quot;Sample_Name&quot;</span>,</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>                     <span class="at">only.pos =</span> <span class="cn">TRUE</span>,</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>                     <span class="at">min.diff.pct =</span> <span class="fl">0.25</span>,</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>                     <span class="at">min.pct =</span> <span class="fl">0.25</span>,</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>                     <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(fcm)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       naive_p_val naive_avg_log2FC naive_pct.1 naive_pct.2 naive_p_val_adj
GNLY  0.000000e+00         4.823298       0.947       0.024    0.000000e+00
GZMA 8.973490e-245         3.778985       0.835       0.041   1.736101e-240
PRF1 1.103550e-125         2.671046       0.453       0.021   2.135038e-121
NKG7 1.504853e-212         3.572770       0.823       0.060   2.911439e-208
CCL5 9.363098e-161         3.319940       0.852       0.120   1.811479e-156
CST7 5.390018e-149         3.014122       0.634       0.053   1.042807e-144
     treated_p_val treated_avg_log2FC treated_pct.1 treated_pct.2
GNLY  0.000000e+00           5.060457         0.961         0.055
GZMA  0.000000e+00           3.813904         0.976         0.082
PRF1  0.000000e+00           3.855625         0.907         0.068
NKG7 8.396253e-261           3.398566         0.982         0.150
CCL5 3.653151e-259           4.133251         0.892         0.108
CST7 6.882951e-236           3.135928         0.759         0.068
     treated_p_val_adj      max_pval minimump_p_val
GNLY      0.000000e+00  0.000000e+00   0.000000e+00
GZMA      0.000000e+00 8.973490e-245   0.000000e+00
PRF1      0.000000e+00 1.103550e-125   0.000000e+00
NKG7     1.624423e-256 1.504853e-212  1.679251e-260
CCL5     7.067750e-255 9.363098e-161  7.306301e-259
CST7     1.331645e-231 5.390018e-149  1.376590e-235</code></pre>
</div>
</div>
<p>We also get a table with information on marker genes. The statistics are computed for each group (naïve vs. treated), and the last two columns are the combined p-value across the two groups.</p>
<ul>
<li><p><code>condition_p_val</code> P-value before multiple testing.</p></li>
<li><p><code>condition_avg_log2FC</code> Log2 fold change on the average of gene expression between one cluster and the rest of the clusters combined.</p></li>
<li><p><code>condition_pct.1</code> Percentage of cells expressing the gene in one group.</p></li>
<li><p><code>condition_pct.2</code> Percentage of cells expressing the gene in another group.</p></li>
<li><p><code>condition_p_val_adj</code> P-value with multiple testing correction.</p></li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The more significant the difference between <code>pct.1</code> and <code>pct.2</code>, the higher the confidence for the marker gene. If <code>pct.1</code> is 0.9 and <code>pct.2</code> is 0.8, it is not as interesting as if <code>pct.2</code> is 0.1.</p>
<p>Marker genes express in a specific cell type. If <code>pct.1</code> is low, such as 0.3, it is not as interesting as if it is 0.9.</p>
</div>
</div>
<p>We can visualize the marker gene expression by <code>FeaturePlot</code> and <code>VlnPlot</code>.</p>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-6_6272c0c99964e6d32c21746da8794f3d">
<div class="sourceCode" id="cb8"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(<span class="at">object =</span> integrate.filtered.rna, </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> <span class="fu">c</span>(<span class="st">&quot;GNLY&quot;</span>, <span class="st">&quot;GZMA&quot;</span>),</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">order =</span> <span class="cn">TRUE</span>,</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">min.cutoff =</span> <span class="st">&#39;q10&#39;</span>, </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">label =</span> <span class="cn">TRUE</span>,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">repel =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<div class="cell-output-display">
<p><img src="06_marker_identification_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="672" /></p>
</div>
</div>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-7_3f664a1455c2c9c398e0bbabb0c4344d">
<div class="sourceCode" id="cb9"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(<span class="at">object =</span> integrate.filtered.rna, </span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">features =</span> <span class="fu">c</span>(<span class="st">&quot;GNLY&quot;</span>, <span class="st">&quot;GZMA&quot;</span>),</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">stack =</span> T, <span class="at">flip =</span> T)</span></code></pre></div>
<div class="cell-output-display">
<p><img src="06_marker_identification_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672" /></p>
</div>
</div>
</section>
<section id="findmarkers" class="level2">
<h2>1.3 <code>FindMarkers</code></h2>
<p>We can use the <code>FindMarkers</code> to determine the differentially expressed genes between two clusters.</p>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-8_03d8e9e68804194fcbbb427d46f4db9f">
<div class="sourceCode" id="cb10"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>example_findmarkers <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(integrate.filtered.rna,</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">assay =</span> <span class="st">&quot;RNA&quot;</span>,</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">ident.1 =</span> <span class="dv">1</span>,</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>                          <span class="at">ident.2 =</span> <span class="dv">2</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="st">&quot;gene&quot;</span>)</span></code></pre></div>
</div>
</section>
</section>
<section id="automatic-annotation" class="level1">
<h1>2. Automatic annotation</h1>
<p>So far, we already look at the marker genes based on knowledge (manual annotation) and between clusters by differential expression analysis. The manual annotation usually takes time, and we might need to go back and forth to annotate the clusters. Another way to annotate the clusters is through the automatic annotation.</p>
<section id="reference-based" class="level2">
<h2>2.1 Reference-based</h2>
<p>We look at the publicly available dataset with annotation and build a reference set based on its information. Then we can “transfer” the annotation from the reference to the query. Seurat provides the interface to connect with <code>Azimuth</code>.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Because the annotation information is based on the reference set, the reference ideally is similar to the query. They may come from the same tissue, same organism, similar experimental procedure, etc.</p>
</div>
</div>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-9_0e43b46fc4bd4c7933d3c01f2c7ba26c">
<div class="sourceCode" id="cb11"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>integrate.filtered.rna <span class="ot">&lt;-</span> Azimuth<span class="sc">::</span><span class="fu">RunAzimuth</span>(integrate.filtered.rna,</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>                                              <span class="at">reference =</span> <span class="st">&quot;pbmcref&quot;</span>,</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>                                              <span class="at">verbose =</span> F)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An AzimuthData object - reference version: 1.0.0 
Contains 3 meta.data field(s) to transfer.</code></pre>
</div>
</div>
<p>We can look at the automatic annotation on UMAP.</p>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-10_a045b9c876c53493393d5949e42fe1de">
<div class="sourceCode" id="cb13"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(integrate.filtered.rna, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, </span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by =</span> <span class="st">&quot;predicted.celltype.l1&quot;</span>, </span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>              <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">label.size =</span> <span class="dv">3</span> ,<span class="at">repel =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> </span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>()</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(integrate.filtered.rna, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, </span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>              <span class="at">group.by =</span> <span class="st">&quot;integrated_snn_res.0.6&quot;</span>, </span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>              <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">label.size =</span> <span class="dv">3</span> ,<span class="at">repel =</span> <span class="cn">TRUE</span>) <span class="sc">+</span> </span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NoLegend</span>()</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>p1<span class="sc">|</span>p2</span></code></pre></div>
<div class="cell-output-display">
<p><img src="06_marker_identification_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672" /></p>
</div>
</div>
</section>
<section id="marker-based" class="level2">
<h2>2.1 Marker-based</h2>
<p>We use ScType (<a href="https://github.com/IanevskiAleksandr/sc-type/blob/master/README.md" class="uri">https://github.com/IanevskiAleksandr/sc-type/blob/master/README.md</a>) to annotate cell population based on known immune markers.</p>
<p>ScType also has an open source portal: <a href="https://sctype.app/" class="uri">https://sctype.app/</a>.</p>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-11_1f8bb527c0b7b52fe50670d1b473cb6e">
<div class="sourceCode" id="cb14"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/gene_sets_prepare.R&quot;</span>); <span class="fu">source</span>(<span class="st">&quot;https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/R/sctype_score_.R&quot;</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="co"># get cell-type-specific gene sets from our in-built database (DB)</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co"># list downloaded from https://raw.githubusercontent.com/IanevskiAleksandr/sc-type/master/ScTypeDB_short.xlsx</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="co"># e.g. Immune system, Liver, Pancreas, Kidney, Eye, Brain</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>gs_list <span class="ot">&lt;-</span>  <span class="fu">gene_sets_prepare</span>(<span class="st">&quot;data/ScTypeDB_short.xlsx&quot;</span>, <span class="st">&quot;Immune system&quot;</span>) </span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="co"># assign cell types</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="co"># </span><span class="al">NOTE</span><span class="co">: scRNAseqData parameter should correspond to your input scRNA-seq matrix. </span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a><span class="co"># In case Seurat is used, it is either pbmc[[&quot;RNA&quot;]]@scale.data (default), pbmc[[&quot;SCT&quot;]]@scale.data, in case sctransform is used for normalization,</span></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="co"># or pbmc[[&quot;integrated&quot;]]@scale.data, in case a joint analysis of multiple single-cell datasets is performed.</span></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>es.max <span class="ot">&lt;-</span>  <span class="fu">sctype_score</span>(<span class="at">scRNAseqData =</span> integrate.filtered.rna<span class="sc">@</span>assays[[<span class="st">&quot;integrated&quot;</span>]]<span class="sc">@</span>scale.data, </span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>                        <span class="at">scaled =</span> <span class="cn">TRUE</span>, </span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>                        <span class="at">gs =</span> gs_list<span class="sc">$</span>gs_positive, </span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>                        <span class="at">gs2 =</span> gs_list<span class="sc">$</span>gs_negative)</span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a><span class="co"># merge by cluster</span></span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>cL_resutls <span class="ot">=</span> <span class="fu">do.call</span>(<span class="st">&quot;rbind&quot;</span>, <span class="fu">lapply</span>(<span class="fu">unique</span>(integrate.filtered.rna<span class="sc">@</span>meta.data<span class="sc">$</span>integrated_snn_res.<span class="fl">0.6</span>), <span class="cf">function</span>(cl){</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>    es.max.cl <span class="ot">=</span> <span class="fu">sort</span>(<span class="fu">rowSums</span>(es.max[ ,<span class="fu">rownames</span>(integrate.filtered.rna<span class="sc">@</span>meta.data[integrate.filtered.rna<span class="sc">@</span>meta.data<span class="sc">$</span>integrated_snn_res.<span class="fl">0.6</span><span class="sc">==</span>cl, ])]), <span class="at">decreasing =</span> <span class="sc">!</span><span class="dv">0</span>)</span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">head</span>(<span class="fu">data.frame</span>(<span class="at">cluster =</span> cl, <span class="at">type =</span> <span class="fu">names</span>(es.max.cl), <span class="at">scores =</span> es.max.cl, <span class="at">ncells =</span> <span class="fu">sum</span>(integrate.filtered.rna<span class="sc">@</span>meta.data<span class="sc">$</span>integrate.filtered.rna<span class="sc">==</span>cl)), <span class="dv">10</span>)</span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a>}))</span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a>sctype_scores <span class="ot">=</span> cL_resutls <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span> <span class="fu">top_n</span>(<span class="at">n =</span> <span class="dv">1</span>, <span class="at">wt =</span> scores)  </span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true" tabindex="-1"></a><span class="co"># set low-confident (low ScType score) clusters to &quot;unknown&quot;</span></span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true" tabindex="-1"></a>sctype_scores<span class="sc">$</span>type[<span class="fu">as.numeric</span>(<span class="fu">as.character</span>(sctype_scores<span class="sc">$</span>scores)) <span class="sc">&lt;</span> sctype_scores<span class="sc">$</span>ncells<span class="sc">/</span><span class="dv">4</span>] <span class="ot">=</span> <span class="st">&quot;Unknown&quot;</span></span></code></pre></div>
</div>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-12_1f19449d28fc5312755e87b8c0c27934">
<div class="sourceCode" id="cb15"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>integrate.filtered.rna<span class="sc">@</span>meta.data<span class="sc">$</span>customclassif <span class="ot">=</span> <span class="st">&quot;&quot;</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(j <span class="cf">in</span> <span class="fu">unique</span>(sctype_scores<span class="sc">$</span>cluster)){</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  cl_type <span class="ot">=</span> sctype_scores[sctype_scores<span class="sc">$</span>cluster<span class="sc">==</span>j,]; </span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  integrate.filtered.rna<span class="sc">@</span>meta.data<span class="sc">$</span>customclassif[integrate.filtered.rna<span class="sc">@</span>meta.data<span class="sc">$</span>integrated_snn_res.<span class="fl">0.6</span> <span class="sc">==</span> j] <span class="ot">=</span> <span class="fu">as.character</span>(cl_type<span class="sc">$</span>type[<span class="dv">1</span>])</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(integrate.filtered.rna, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">repel =</span> <span class="cn">TRUE</span>, <span class="at">group.by =</span> <span class="st">&#39;customclassif&#39;</span>)   </span></code></pre></div>
<div class="cell-output-display">
<p><img src="06_marker_identification_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672" /></p>
</div>
</div>
</section>
</section>
<section id="save" class="level1">
<h1>3. Save</h1>
<p>Save the files into RData for documentation.</p>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-13_500824c467b35a7c2a1afcc4fa6c6fae">
<div class="sourceCode" id="cb16"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save(integrate.filtered.rna, expMat, file = &quot;data/intermediate_data/chapter6.RData&quot;)</span></span></code></pre></div>
</div>
<div id="quarto-navigation-envelope" class="hidden">
<p><span class="hidden" data-render-id="quarto-int-sidebar-title">Single Cell Boot Camp: Analyzing Single Cell RNA-seq with R</span> <span class="hidden" data-render-id="quarto-int-navbar-title">Chapter 6 Marker Identification</span> <span class="hidden" data-render-id="quarto-int-next">Chapter 8 Enrichment Analysis</span> <span class="hidden" data-render-id="quarto-int-prev">Chapter 5 Clustering</span> <span class="hidden" data-render-id="quarto-int-sidebar:/index.html">Introduction</span> <span class="hidden" data-render-id="quarto-int-sidebar:/setup.html">Setting up</span> <span class="hidden" data-render-id="quarto-int-sidebar:/precourse_preparation_R.html">Precourse Preparation - R</span> <span class="hidden" data-render-id="quarto-int-sidebar:/precourse_preparation_theory.html">Precourse Preparation - Theory</span> <span class="hidden" data-render-id="quarto-int-sidebar:/01_data_import.html">Chapter 1 Data Import</span> <span class="hidden" data-render-id="quarto-int-sidebar:/02_quality_control.html">Chapter 2 Quality Control</span> <span class="hidden" data-render-id="quarto-int-sidebar:/03_normalization.html">Chapter 3 Normalization and PCA</span> <span class="hidden" data-render-id="quarto-int-sidebar:/04_integration.html">Chapter 4 Integration</span> <span class="hidden" data-render-id="quarto-int-sidebar:/05_clustering.html">Chapter 5 Clustering</span> <span class="hidden" data-render-id="quarto-int-sidebar:/06_marker_identification.html">Chapter 6 Marker Identification</span> <span class="hidden" data-render-id="quarto-int-sidebar:/08_enrichment_analysis.html">Chapter 8 Enrichment Analysis</span> <span class="hidden" data-render-id="quarto-int-sidebar:/09_combining.html">Chapter 9 Combine scRNAseq and AbSeq</span> <span class="hidden" data-render-id="quarto-int-sidebar:/supplementary_targeted_analysis.html">Supplementary: Data import and quality control for targeted assays</span> <span class="hidden" data-render-id="quarto-int-sidebar:/sessioninfo.html">SessionInfo</span> <span class="hidden" data-render-id="quarto-breadcrumbs-d8c230701586fbacdd4ae74c76d4af5d">Chapter 6 Marker Identification</span></p>
</div>
<div id="quarto-meta-markdown" class="hidden">
<p><span class="hidden" data-render-id="quarto-metatitle">Chapter 6 Marker Identification</span> <span class="hidden" data-render-id="quarto-twittercardtitle">Chapter 6 Marker Identification</span> <span class="hidden" data-render-id="quarto-ogcardtitle">Chapter 6 Marker Identification</span> <span class="hidden" data-render-id="quarto-twittercarddesc"></span> <span class="hidden" data-render-id="quarto-ogcardddesc"></span></p>
</div>
</section>

</main> <!-- /main -->
<script id = "quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->

</body>

</html>