---
title: "scRepertoire"
execute:
  warning: false
  message: false
  eval: false
format:
  html:
    embed-resources: true
---

## Installation

This tutorial is to show how to import VDJ output from BD Rhapsody primary pipeline into scRepertorie.

Please install the latest version of scRepertoire from the master branch of github repo. It has to be the latest version to have the functions running.

```{r}
devtools::install_github("ncborcherding/scRepertoire")
```

## Library

```{r}
library(tidyverse)
library(scRepertoire)
library(data.table)
library(SeuratDisk) 
```

## 1. Read in data

```{r}
# read in seurat object
## Note: When analysing a subset of cells, eg. QC to keep high quality cells, you need to create a seurat object only having that subset of cells.
seuratObj <- readRDS("Seurat.rds")

# read in contig file
contigs <- fread("VDJ_Dominant_Contigs_AIRR.tsv") %>% 
  rename(chain = locus,
         barcode = cell_id,
         reads = consensus_count,
         v_gene = v_call,
         d_gene = d_call,
         j_gene = j_call,
         c_gene = c_call,
         cdr3_nt = cdr3,
         cdr3 = cdr3_aa)

# create a contig list
contig.list <- createHTOContigList(contigs,
                                  seuratObj,
                                  group.by = "orig.ident")

# insert sample names
combined.TCR <- combineTCR(contig.list, samples = c("orig.ident"))
```

**Below code is only for experiments with SMK.**

```{r}
# create a contig list
contig.list <- createHTOContigList(contigs,
                                  seuratObj,
                                  group.by = "Sample_Name")

# insert sample names
combined.TCR <- combineTCR(contig.list, samples = c(""))
```

## 2. scRepertoire

From here onward, you can refer to the tutorial to visualize VDJ data <https://www.borch.dev/uploads/screpertoire/articles/clonal_visualizations>

```{r}
clonalQuant(combined.TCR)
```

From here onward, you can refer to this tutorial to visualize data on dimentionality graphs. <https://www.borch.dev/uploads/screpertoire/articles/attaching_sc>

```{r}
# update cell id names
colnames(seuratObj) <- paste(seuratObj$Sample_Name, colnames(seuratObj), sep = "_")

# convert to sce and add TCR information
sce <- Seurat::as.SingleCellExperiment(seuratObj)

sce <- combineExpression(combined.TCR, 
                         sce,
                         cloneCall = "gene",
                         group.by = "sample",
                         proportion = T)

## Optional: convert back to seurat for visualization
# seuratObj_convertBack <- as.Seurat(sce, counts = "counts", data = NULL) 
```

## Reference

Github page: <https://github.com/BorchLab/scRepertoire>

Vignette: <https://www.borch.dev/uploads/screpertoire/>
