---
title: "Use GPTCelltype to annotate cell types"
execute:
  warning: false
  message: false
format:
  html:
    embed-resources: true
---

## Introduction
Github repository: https://github.com/Winnie09/GPTCelltype

GPTCelltype is a package that uses OpenAI's GPT-3 and GPT-4 models to annotate cell types based on marker genes. The package is designed to work with Seurat objects and can be used to annotate cell types in single-cell RNA-seq data.

## Installation
```{r}
#| eval: FALSE
install.packages("openai")
remotes::install_github("Winnie09/GPTCelltype")
```

## Load libraries
```{r}
library(tidyverse)
library(Seurat)
library(GPTCelltype)
```

## Example

1. Import example data
```{r}
load("data/chapter5.RData")
```

2. Annotation of cell types
```{r}
DimPlot(filtered, group.by = "predicted.celltype.l1", label = T)
```

3. Use annotation
```{r}
# IMPORTANT! Assign your OpenAI API key. See Vignette for details
Sys.setenv(OPENAI_API_KEY = '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')

# Assume you have already run the Seurat pipeline https://satijalab.org/seurat/
# "obj" is the Seurat object; "markers" is the output from FindAllMarkers(obj)
# Cell type annotation by GPT-4
res <- gptcelltype(treated.markers, model = 'gpt-4')

# Assign cell type annotation back to Seurat object
filtered@meta.data$celltype <- as.factor(res[as.character(Idents(filtered))])

# Visualize cell type annotation on UMAP
DimPlot(obj,group.by='celltype')
```

