<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.340">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Chapter 6 Marker Identification</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script><link rel="stylesheet" href="styles.css">
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./06_marker_identification.html">Chapter 6 Marker Identification</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Single Cell Boot Camp: Analyzing Single Cell RNA-seq with R</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./about.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Setting up</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./precourse_preparation_R.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Precourse Preparation - R</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./precourse_preparation_theory.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Precourse Preparation - Theory</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_data_import.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 1 Data Import</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 2 Quality Control</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03_normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 3 Normalization and PCA</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04_integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 4 Integration</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 5 Clustering</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06_marker_identification.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Chapter 6 Marker Identification</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07_differential_expression_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 7 Differential Expression Analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08_enrichment_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 8 Enrichment Analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09_combining.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 9 Combine scRNAseq and AbSeq</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./sessioninfo.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">SessionInfo</span></a>
  </div>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li><a href="#load-rdata" id="toc-load-rdata" class="nav-link active" data-scroll-target="#load-rdata">Load RData</a></li>
  <li>
<a href="#find-marker-genes" id="toc-find-marker-genes" class="nav-link" data-scroll-target="#find-marker-genes">1. Find marker genes</a>
  <ul class="collapse">
<li><a href="#findallmarkers" id="toc-findallmarkers" class="nav-link" data-scroll-target="#findallmarkers">1.1 FindAllMarkers</a></li>
  <li><a href="#findconservedmarkers" id="toc-findconservedmarkers" class="nav-link" data-scroll-target="#findconservedmarkers">1.2 FindConservedMarkers</a></li>
  <li><a href="#findmarkers" id="toc-findmarkers" class="nav-link" data-scroll-target="#findmarkers">1.3 FindMarkers</a></li>
  </ul>
</li>
  <li>
<a href="#automatic-annotation" id="toc-automatic-annotation" class="nav-link" data-scroll-target="#automatic-annotation">2. Automatic annotation</a>
  <ul class="collapse">
<li><a href="#reference-based" id="toc-reference-based" class="nav-link" data-scroll-target="#reference-based">2.1 Reference-based</a></li>
  <li><a href="#marker-based" id="toc-marker-based" class="nav-link" data-scroll-target="#marker-based">2.1 Marker-based</a></li>
  </ul>
</li>
  <li><a href="#save" id="toc-save" class="nav-link" data-scroll-target="#save">3. Save</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">Chapter 6 Marker Identification</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-1_e3a286f08524b0e372da9479ee573b3d">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># libraries for this chapter</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/celldex">celldex</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/SingleR">SingleR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/FelixTheStudent/cellpypes">cellpypes</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="load-rdata" class="level1"><h1>Load RData</h1>
<p>Load the RData which contain all files from previous chapter.</p>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-2_38dd3572e2382bc4cda1f5fa67a0e553">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/intermediate_data/chapter5.RData"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="find-marker-genes" class="level1"><h1>1. Find marker genes</h1>
<section id="findallmarkers" class="level2"><h2 class="anchored" data-anchor-id="findallmarkers">1.1 FindAllMarkers</h2>
<p>This type of analysis is typically for evaluating a single sample group/condition. With the <code>FindAllMarkers</code> function we are comparing each cluster against all other clusters to identify potential marker genes. The cells in each cluster are treated as replicates, and essentially a differential expression analysis is performed with some statistical test. By default, the test is Wilcoxon Rank Sum test.</p>
<p>The <code>FindAllMarkers</code> function has three arguments to determine if a gene is a marker gene.</p>
<p><code>logfc.threshold</code> Limit testing to genes that show, on average, at least an X-fold difference (log-scale) between the two groups of cells. The default is 0.25. Increasing logfc.threshold speeds up the function but can miss weaker signals.</p>
<p><code>min.pct</code> Only test genes that are detected in a minimum fraction of min.pct cells in either of the two populations. It is meant to speed up the function by not testing infrequently expressed genes. The default is 0.1.</p>
<p><code>only.pos</code> Only return positive markers (FALSE by default)</p>
<p>We can combine these arguments. We can change the arguments depending on how stringent we want to be. Here we will use the default settings.</p>
<p>We look at the marker genes in the clusters in sample “activated”.</p>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-3_26155b8394f9ef7254e0d5271da9ad5a">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">treated.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html">subset</a></span><span class="op">(</span><span class="va">integrate.filtered.rna</span>, subset <span class="op">=</span> <span class="va">Sample_Name</span> <span class="op">==</span> <span class="st">"treated"</span><span class="op">)</span>, </span>
<span>                                    assay <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>                                    only.pos <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                    min.pct <span class="op">=</span> <span class="fl">0.25</span>, </span>
<span>                                    logfc.threshold <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s look at the top 2 significant marker genes in each cluster.</p>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-4_d6a5768d4bc4ed9c8792446e4bae4831">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">treated.markers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_max</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">2</span>, order_by <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 26 × 7
# Groups:   cluster [13]
       p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene      
       &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt;     
 1 9.89e-158       2.63 0.744 0.212 1.91e-153 0       MAL       
 2 3.62e-104       2.10 0.734 0.379 7.01e-100 0       PCED1B-AS1
 3 0               5.06 0.961 0.055 0         1       GNLY      
 4 3.65e-259       4.13 0.892 0.108 7.07e-255 1       CCL5      
 5 5.52e- 12       1.07 0.75  0.515 1.07e-  7 2       HSPD1     
 6 6.86e- 17       1.00 0.957 0.787 1.33e- 12 2       NPM1      
 7 5.53e- 82       2.41 0.594 0.119 1.07e- 77 3       RGS1      
 8 5.81e- 25       1.82 0.304 0.09  1.12e- 20 3       FOXP3     
 9 1.24e-122       2.25 0.916 0.276 2.39e-118 4       LTA       
10 1.65e-159       1.97 0.941 0.225 3.20e-155 4       MIR155HG  
# ℹ 16 more rows</code></pre>
</div>
</div>
<p>We get a table with the information on marker genes.</p>
<p><code>p_val</code> p-value before multiple testing.</p>
<p><code>avg_log2FC</code> log2 fold change on the average of gene expression between one cluster and the rest of cluster combined.</p>
<p><code>pct.1</code> Percentage of cells express the gene in one group. Here is one cluster.</p>
<p><code>pct.2</code> Percentage of cells express the gene in another group. Here is the rest of cluster combined.</p>
<p><code>p_val_adj</code> p-value with multiple testing correction.</p>
</section><section id="findconservedmarkers" class="level2"><h2 class="anchored" data-anchor-id="findconservedmarkers">1.2 FindConservedMarkers</h2>
<p>We have two samples, naïve and treated. We can find conserved markers between the conditions with the function, <code>FindConservedMarkers</code>.</p>
<p>This function internally separates out cells by sample group/condition, then performs differential gene expression testing for a single specified cluster against all other clusters (or a second cluster, if specified).</p>
<p>Gene-level p-values are computed for each condition and then combined across groups using meta-analysis methods from the <code>MetaDE</code> R package.</p>
<p>Many of the arguments in <code>FindConservedMarkers</code> are shared with <code>FindAllMarkers</code>. Here, we list some additional arguments when using <code>FindConservedMarkers</code>:</p>
<p><code>ident.1</code> This function only evaluates one cluster at a time; here, you would specify the cluster of interest. <code>grouping.var</code> The variable (column header) in your metadata specifies the separation of cells into groups.</p>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-5_9342c7f74876e97e712d02fcf809915f">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">fcm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindConservedMarkers.html">FindConservedMarkers</a></span><span class="op">(</span><span class="va">integrate.filtered.rna</span>,</span>
<span>                     assay <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>                     ident.1 <span class="op">=</span> <span class="st">"1"</span>,</span>
<span>                     grouping.var <span class="op">=</span> <span class="st">"Sample_Name"</span>,</span>
<span>                     only.pos <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     min.diff.pct <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>                     min.pct <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>                     logfc.threshold <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">fcm</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       naive_p_val naive_avg_log2FC naive_pct.1 naive_pct.2 naive_p_val_adj
GNLY  0.000000e+00         4.823298       0.947       0.024    0.000000e+00
GZMA 8.973490e-245         3.778985       0.835       0.041   1.736101e-240
PRF1 1.103550e-125         2.671046       0.453       0.021   2.135038e-121
NKG7 1.504853e-212         3.572770       0.823       0.060   2.911439e-208
CCL5 9.363098e-161         3.319940       0.852       0.120   1.811479e-156
CST7 5.390018e-149         3.014122       0.634       0.053   1.042807e-144
     treated_p_val treated_avg_log2FC treated_pct.1 treated_pct.2
GNLY  0.000000e+00           5.060457         0.961         0.055
GZMA  0.000000e+00           3.813904         0.976         0.082
PRF1  0.000000e+00           3.855625         0.907         0.068
NKG7 8.396253e-261           3.398566         0.982         0.150
CCL5 3.653151e-259           4.133251         0.892         0.108
CST7 6.882951e-236           3.135928         0.759         0.068
     treated_p_val_adj      max_pval minimump_p_val
GNLY      0.000000e+00  0.000000e+00   0.000000e+00
GZMA      0.000000e+00 8.973490e-245   0.000000e+00
PRF1      0.000000e+00 1.103550e-125   0.000000e+00
NKG7     1.624423e-256 1.504853e-212  1.679251e-260
CCL5     7.067750e-255 9.363098e-161  7.306301e-259
CST7     1.331645e-231 5.390018e-149  1.376590e-235</code></pre>
</div>
</div>
<p>We also get a table with information on marker genes. The statistics are computed for each group (naïve vs.&nbsp;treated), and the last two columns are the combined p-value across the two groups.</p>
<p><code>condition_p_val</code> p-value before multiple testing.</p>
<p><code>condition_avg_log2FC</code> log2 fold change on the average of gene expression between one cluster and the rest of the clusters combined.</p>
<p><code>condition_pct.1</code> Percentage of cells expressing the gene in one group.</p>
<p><code>condition_pct.2</code> Percentage of cells expressing the gene in another group.</p>
<p><code>condition_p_val_adj</code> p-value with multiple testing correction.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The more significant the difference between <code>pct.1</code> and <code>pct.2</code>, the higher the confidence for the marker gene. If <code>pct.1</code> is 0.9 and <code>pct.2</code> is 0.8, it is not as interesting as if <code>pct.2</code> is 0.1.</p>
<p>Marker genes express in a specific cell type. If <code>pct.1</code> is low, such as 0.3, it is not as interesting as if it is 0.9.</p>
</div>
</div>
<p>We can visualize the marker gene expression by <code>FeaturePlot</code> and <code>VlnPlot</code>.</p>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-6_6272c0c99964e6d32c21746da8794f3d">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">integrate.filtered.rna</span>, </span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GNLY"</span>, <span class="st">"GZMA"</span><span class="op">)</span>,</span>
<span>            order <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            min.cutoff <span class="op">=</span> <span class="st">'q10'</span>, </span>
<span>            label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="06_marker_identification_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-7_8bb98b06e6afc1251a5ec0e7141467c1">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">integrate.filtered.rna</span>, </span>
<span>        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"GNLY"</span>, <span class="st">"GZMA"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="06_marker_identification_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section><section id="findmarkers" class="level2"><h2 class="anchored" data-anchor-id="findmarkers">1.3 FindMarkers</h2>
<p>We can use the <code>FindMarkers</code> to determine the differentially expressed genes between two clusters.</p>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-8_03d8e9e68804194fcbbb427d46f4db9f">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">example_findmarkers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">integrate.filtered.rna</span>,</span>
<span>                          assay <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>                          ident.1 <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                          ident.2 <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="st">"gene"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="automatic-annotation" class="level1"><h1>2. Automatic annotation</h1>
<p>So far, we already look at the marker genes based on knowledge (last chapter) and between clusters. The manual annotation usually takes time, and we might need to go back and forth to annotate the clusters. Another way to annotate the clusters is through the automatic annotation.</p>
<section id="reference-based" class="level2"><h2 class="anchored" data-anchor-id="reference-based">2.1 Reference-based</h2>
<p>We look at the publicly available dataset with annotation and build a reference set based on its information. Then we can “transfer” the annotation from the reference to the query. Seurat provides the interface to connect with Azimuth.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Because the annotation information is based on the reference set, the reference ideally are similar to the query. They may come from the same tissue, same organism, similar experimental procedure, etc.</p>
</div>
</div>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-9_6c55e00ec0d488fdbf68fdabae10602e">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">integrate.filtered.rna</span> <span class="op">&lt;-</span> <span class="fu">Azimuth</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Azimuth/man/RunAzimuth.html">RunAzimuth</a></span><span class="op">(</span><span class="va">integrate.filtered.rna</span>, reference <span class="op">=</span> <span class="st">"pbmcref"</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An AzimuthData object - reference version: 1.0.0 
Contains 3 meta.data field(s) to transfer.</code></pre>
</div>
</div>
<p>We can look at the prediction on UMAP.</p>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-10_13bf7bf08475c6184f85bb52a30ea5eb">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">integrate.filtered.rna</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"predicted.celltype.l1"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, label.size <span class="op">=</span> <span class="fl">3</span> ,repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">integrate.filtered.rna</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, group.by <span class="op">=</span> <span class="st">"integrated_snn_res.0.6"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, label.size <span class="op">=</span> <span class="fl">3</span> ,repel <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span><span class="op">|</span><span class="va">p2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="06_marker_identification_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section><section id="marker-based" class="level2"><h2 class="anchored" data-anchor-id="marker-based">2.1 Marker-based</h2>
<p>The <code>celldex</code> contains several curated reference datasets, mostly assembled from bulk RNA-seq or microarray data of sorted cell types. Here, we will use a reference constructed from Blueprint and ENCODE data (Martens and Stunnenberg 2013; The ENCODE Project Consortium 2012).</p>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-11_7e7ac7accc6207d6ce94d76edbee667d">
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># get reference</span></span>
<span><span class="co"># ref_encode &lt;- BlueprintEncodeData()</span></span>
<span><span class="va">ref_encode</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"data/blueprint_encode_ref.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># perform annotation</span></span>
<span><span class="va">annot_encode</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html">SingleR</a></span><span class="op">(</span>test <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/as.SingleCellExperiment.html">as.SingleCellExperiment</a></span><span class="op">(</span><span class="va">integrate.filtered.rna</span><span class="op">)</span>, ref <span class="op">=</span> <span class="va">ref_encode</span>, labels <span class="op">=</span> <span class="va">ref_encode</span><span class="op">$</span><span class="va">label.main</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get labels into Seurat object</span></span>
<span><span class="va">integrate.filtered.rna</span><span class="op">$</span><span class="va">annot_encode</span> <span class="op">&lt;-</span> <span class="va">annot_encode</span><span class="op">$</span><span class="va">labels</span></span>
<span></span>
<span><span class="co"># Visualizae</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">integrate.filtered.rna</span>, group.by <span class="op">=</span> <span class="st">"annot_encode"</span>, label <span class="op">=</span> <span class="cn">T</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="06_marker_identification_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section></section><section id="save" class="level1"><h1>3. Save</h1>
<p>Save the files into RData for documentation.</p>
<div class="cell" data-hash="06_marker_identification_cache/html/unnamed-chunk-12_bf5f1dd5b098c29d2ff37716fca1fc00">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># save(integrate.filtered.rna, expMat, file = "data/intermediate_data/chapter6.RData")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>