<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.340">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Chapter 8 Enrichment Analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script><link rel="stylesheet" href="styles.css">
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">
      Single Cell Boot Camp: Analyzing Single Cell RNA-seq with R
      </li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Single Cell Boot Camp: Analyzing Single Cell RNA-seq with R</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./about.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Setting up</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./precourse_preparation_R.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Precourse Preparation - R</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./precourse_preparation_theory.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Precourse Preparation - Theory</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_data_import.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 1 Data Import</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 2 Quality Control</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03_normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 3 Normalization and PCA</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04_integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 4 Integration</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 5 Clustering</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06_marker_identification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 6 Marker Identification</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07_differential_expression_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 7 Differential Expression Analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08_enrichment_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 8 Enrichment Analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09_combining.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 9 Combine scRNAseq and AbSeq</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./sessioninfo.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">SessionInfo</span></a>
  </div>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li><a href="#load-rdata" id="toc-load-rdata" class="nav-link active" data-scroll-target="#load-rdata">Load RData</a></li>
  <li><a href="#get-entrez-id" id="toc-get-entrez-id" class="nav-link" data-scroll-target="#get-entrez-id">1. Get Entrez ID</a></li>
  <li><a href="#go-over-representation-analysis" id="toc-go-over-representation-analysis" class="nav-link" data-scroll-target="#go-over-representation-analysis">2. GO over-representation analysis</a></li>
  <li><a href="#gene-set-enrichment-analysis" id="toc-gene-set-enrichment-analysis" class="nav-link" data-scroll-target="#gene-set-enrichment-analysis">3. Gene set enrichment analysis</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">Chapter 8 Enrichment Analysis</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="cell" data-hash="08_enrichment_analysis_exercise_cache/html/unnamed-chunk-1_75c1559f874c2bdbc1b560c428929001">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># libraries for this chapter</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/celldex">celldex</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/SingleR">SingleR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/">clusterProfiler</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/AnnotationDbi">AnnotationDbi</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yulab-smu.top/biomedical-knowledge-mining-book/">enrichplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/grimbough/biomaRt">biomaRt</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="load-rdata" class="level1"><h1>Load RData</h1>
<p>Load the RData which contains all files from the previous chapter.</p>
<div class="cell" data-hash="08_enrichment_analysis_exercise_cache/html/unnamed-chunk-2_275712679b8ecc02ffa3b653152d57e8">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/intermediate_data/chapter7.RData"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This chapter uses the differentially expressed genes in CD4+ T cells from last chapter as an example.</p>
</section><section id="get-entrez-id" class="level1"><h1>1. Get Entrez ID</h1>
<p><code>clusterProfiler</code> prefers the input of Entrez ID. We use <code>biomaRt</code> package to convert the gene symbol into Entrez ID.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p><code>biomaRt</code> is handy to get information about features. You can use it to get gene symbols, genomic coordinates, transcript sequences, etc. You can explore what data is stored in <code>biomaRt</code> with <code>listAttributes</code>.</p>
</div>
</div>
<p>The differentially expressed genes in CD4+ T cells between naïve and treated PBMCs are stored in the matrix <code>cd4t</code>. The gene symbol in this list can be converted into Entrez ID by <code>biomaRt</code>.</p>
<p>Firstly, connect to the Ensembl database and specify we want to retrieve information from the dataset of Homo Sapien.</p>
<div class="cell" data-hash="08_enrichment_analysis_exercise_cache/html/unnamed-chunk-3_f76a05431a20be23c9705edb19a0d9e6">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mart</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span><span class="op">(</span>biomart <span class="op">=</span> <span class="st">"ensembl"</span>, dataset <span class="op">=</span> <span class="st">"hsapiens_gene_ensembl"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><p><code>biomart</code> BioMart database connect to. Here, we want to connect to <code>ensembl</code> database.</p></li>
<li><p><code>dataset</code> Dataset to use. Here specify <code>hsapiens_gene_ensembl</code>.</p></li>
</ul>
<p>Next, we can give a list of gene symbols and retrieve the Entrez ID through <code>getBM</code>.</p>
<div class="cell" data-hash="08_enrichment_analysis_exercise_cache/html/unnamed-chunk-4_5f00c94090226215f4a3b74bcb442710">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span><span class="op">(</span>attributes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"entrezgene_id"</span>, <span class="st">"external_gene_name"</span><span class="op">)</span>,</span>
<span>                 filters <span class="op">=</span> <span class="st">"external_gene_name"</span>, </span>
<span>                 values <span class="op">=</span> <span class="va">cd4t</span><span class="op">$</span><span class="va">gene</span>,</span>
<span>                 mart <span class="op">=</span> <span class="va">mart</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li><p><code>attributes</code> Attributes to retrieve. We want to have the gene symbols and Entrez ID in the results.</p></li>
<li><p><code>filters</code> Which filter to use to filter the dataset.</p></li>
<li><p><code>values</code> A list of features as a query.</p></li>
<li><p><code>mart</code> Object of class Mart, created with the <code>useMart</code> function.</p></li>
</ul>
<p>Not every gene has the Entrez ID. We keep only genes that have a matched Entrez ID.</p>
<div class="cell" data-hash="08_enrichment_analysis_exercise_cache/html/unnamed-chunk-5_49aa3f7a26525b056f715d83e66a00e0">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cd4t</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="va">cd4t</span>, <span class="va">results</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"gene"</span> <span class="op">=</span> <span class="st">"external_gene_name"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">entrezgene_id</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="go-over-representation-analysis" class="level1"><h1>2. GO over-representation analysis</h1>
<p>We use the <code>enrichGO</code> to get the enriched GO terms in the list of DEGs.</p>
<div class="cell" data-hash="08_enrichment_analysis_exercise_cache/html/unnamed-chunk-6_242bc3447a996ae88268beae4f46ef73">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cd4t.sig</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">cd4t</span>, <span class="va">p_adj.loc</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">p_adj.loc</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get the enriched GO terms</span></span>
<span><span class="va">ego</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichGO.html">enrichGO</a></span><span class="op">(</span>gene <span class="op">=</span> <span class="va">cd4t.sig</span><span class="op">$</span><span class="va">entrezgene_id</span>,</span>
<span>                keyType <span class="op">=</span> <span class="st">"ENTREZID"</span>,</span>
<span>                OrgDb <span class="op">=</span> <span class="va">org.Hs.eg.db</span>,</span>
<span>                ont <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>                pAdjustMethod <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>                pvalueCutoff <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>                qvalueCutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                readable <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s convert the <code>ego</code> object into a data frame and visualize the results in a bar plot.</p>
<div class="cell" data-hash="08_enrichment_analysis_exercise_cache/html/unnamed-chunk-7_49a8f7736464b79bd39143e09f789e1b">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ego.table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">ego</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ego.table</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   ID                                             Description
GO:0051056 GO:0051056 regulation of small GTPase mediated signal transduction
GO:0007264 GO:0007264               small GTPase mediated signal transduction
GO:0030217 GO:0030217                                  T cell differentiation
GO:0061912 GO:0061912                                     selective autophagy
GO:0030098 GO:0030098                              lymphocyte differentiation
GO:0007265 GO:0007265                         Ras protein signal transduction
           GeneRatio   BgRatio       pvalue     p.adjust       qvalue
GO:0051056   57/1469 298/18614 2.989748e-10 1.653330e-06 1.537674e-06
GO:0007264   76/1469 483/18614 4.702070e-09 1.300122e-05 1.209174e-05
GO:0030217   48/1469 300/18614 1.992165e-06 3.219710e-03 2.994480e-03
GO:0061912   22/1469  93/18614 2.416404e-06 3.219710e-03 2.994480e-03
GO:0030098   61/1469 422/18614 2.911130e-06 3.219710e-03 2.994480e-03
GO:0007265   51/1469 333/18614 3.575447e-06 3.295371e-03 3.064848e-03
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  geneID
GO:0051056                                                                                                                                   RIPOR2/CDC42SE1/RASA3/P2RY10/AKAP13/LPAR6/ITPKB/FGD3/CYTH1/ARHGAP9/ARHGEF18/ARHGEF1/P2RY8/PRAG1/DEF6/DOCK8/ARHGAP45/ITGB1/GPR18/SOS2/OGT/CYTH4/ARHGAP15/STMN3/ARHGAP12/ARHGAP25/ARHGAP21/SPATA13/ARHGEF11/DENND4C/RASA2/IQSEC1/RASGRF2/ARHGAP4/GPR174/NF1/OBSCN/GPR65/SYNGAP1/PLEKHG3/TAGAP/LPAR2/RASAL3/SIPA1/PLCE1/RALGAPA2/KANK1/ARHGAP35/ITGAV/PIK3CG/SIPA1L3/ABL2/ARHGAP32/ARHGEF9/KBTBD6/KBTBD7/MCF2L
GO:0007264 RIPOR2/RASGRP2/CDC42SE1/RASA3/P2RY10/AKAP13/LPAR6/ITPKB/FGD3/GARRE1/CYTH1/ARHGAP9/ARHGEF18/ARHGEF1/P2RY8/PRAG1/DOCK11/DEF6/DOCK8/ARHGAP45/GRAP/ITGB1/CCDC88C/SDCBP/GPR18/SOS2/NUCB2/RAP2B/OGT/CYTH4/SH2D3C/ARHGAP15/STMN3/ARHGAP12/ARHGAP25/ARHGAP21/SPATA13/ARHGEF11/DENND4C/RASA2/SYNJ2BP/RAPGEF6/IQSEC1/RASGRF2/ARHGAP4/GPR174/RALGPS2/NF1/DOCK9/OBSCN/GPR65/SYNGAP1/RAB33A/PLEKHG3/TNK2/TAGAP/LPAR2/RASAL3/NISCH/SIPA1/PLCE1/RALGAPA2/KANK1/ARHGAP35/ITGAV/PIK3CG/RASGEF1B/SIPA1L3/ABL2/ARHGAP32/ARHGEF9/CDH13/RIPOR1/KBTBD6/KBTBD7/MCF2L
GO:0030217                                                                                                                                                                                                                                                   IL7R/ZFP36L2/TCF7/PTPRC/ZFP36L1/CAMK4/IL6R/LEF1/BCL11B/PIK3R1/LY9/ITPKB/TGFBR2/TESPA1/CD4/RORA/KMT2A/GPR183/RHOH/ANXA1/GPR18/SMAD7/STAT5B/SOS2/PIK3CD/GATA3/ZEB1/FOXO3/ZBTB1/CD46/LFNG/CR1/LMBR1L/CTSL/THEMIS/CHD7/TMEM131L/IL23A/RUNX2/SLAMF6/RC3H2/VSIR/ZAP70/ATP7A/METTL3/MR1/CARD11/TOX
GO:0061912                                                                                                                                                                                                                                                                                                                                                                                                         TOMM7/ATM/VPS13D/VPS13C/RETREG1/OGT/RB1CC1/ATG14/KLHL3/RETREG3/UFL1/OPTN/ATG16L1/PINK1/PRKN/GABARAPL1/ULK1/ATG2B/SNX18/BAG3/SNX30/GAA
GO:0030098                                                                                                                                                               IL7R/ZFP36L2/TCF7/PTPRC/ATM/ZFP36L1/CAMK4/IL6R/LEF1/BCL11B/PIK3R1/LY9/ITPKB/TGFBR2/TESPA1/CD4/RORA/KMT2A/DOCK11/GPR183/ITGB1/TOP2B/RHOH/ANXA1/GPR18/EP300/SMAD7/ZBTB7A/IRF2BP2/STAT5B/SOS2/PIK3CD/ST3GAL1/GATA3/ZEB1/FOXO3/ZBTB1/INPP5D/CD46/LFNG/CR1/LMBR1L/PLCL2/CTSL/THEMIS/CHD7/TMEM131L/HDAC5/IL23A/RUNX2/SLAMF6/RC3H2/PTPRJ/VSIR/ZAP70/DNAJB9/ATP7A/METTL3/MR1/CARD11/TOX
GO:0007265                                                                                                                                                                                       RIPOR2/RASGRP2/CDC42SE1/RASA3/P2RY10/AKAP13/LPAR6/ITPKB/GARRE1/CYTH1/ARHGEF18/ARHGEF1/P2RY8/PRAG1/GRAP/ITGB1/SDCBP/GPR18/SOS2/RAP2B/OGT/CYTH4/STMN3/ARHGEF11/DENND4C/RASA2/SYNJ2BP/RAPGEF6/IQSEC1/RASGRF2/ARHGAP4/GPR174/RALGPS2/NF1/GPR65/SYNGAP1/RAB33A/LPAR2/RASAL3/NISCH/PLCE1/KANK1/ARHGAP35/PIK3CG/RASGEF1B/ABL2/CDH13/RIPOR1/KBTBD6/KBTBD7/MCF2L
           Count
GO:0051056    57
GO:0007264    76
GO:0030217    48
GO:0061912    22
GO:0030098    61
GO:0007265    51</code></pre>
</div>
</div>
<ul>
<li><p><code>GeneRatio</code> The gene counts from the input list found in the gene set of a GO term / gene counts of input list.</p></li>
<li><p><code>BgRatio</code> The gene counts of a geneset / size of all of the unique genes in the collection of genesets.</p></li>
</ul>
<div class="cell" data-hash="08_enrichment_analysis_exercise_cache/html/unnamed-chunk-8_2563f0157c1dd676acf85392214fdb25">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="va">ego</span>, showCategory <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="08_enrichment_analysis_exercise_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section><section id="gene-set-enrichment-analysis" class="level1"><h1>3. Gene set enrichment analysis</h1>
<p>All genes can be used as input of gene set enrichment analysis (GSEA). We first rank the genes based on <code>stat</code>.</p>
<div class="cell" data-hash="08_enrichment_analysis_exercise_cache/html/unnamed-chunk-9_e80d45ecf20cf15c7951c4ae3c9baf50">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># order genes based on stat</span></span>
<span><span class="va">cd4t</span> <span class="op">&lt;-</span> <span class="va">cd4t</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">cd4t</span><span class="op">$</span><span class="va">stat</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">gene_list</span> <span class="op">&lt;-</span> <span class="va">cd4t</span><span class="op">$</span><span class="va">stat</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">gene_list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">cd4t</span><span class="op">$</span><span class="va">entrezgene_id</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We use <code>gseGO</code> to perform GSEA and convert the results into a dataframe.</p>
<div class="cell" data-hash="08_enrichment_analysis_exercise_cache/html/unnamed-chunk-10_666d2c2bf68ed065d3ba4512314b2cf7">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># perform GSEA</span></span>
<span><span class="va">gse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/gseGO.html">gseGO</a></span><span class="op">(</span><span class="va">gene_list</span>,</span>
<span>             keyType <span class="op">=</span> <span class="st">"ENTREZID"</span>,</span>
<span>             ont <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>             OrgDb <span class="op">=</span> <span class="st">"org.Hs.eg.db"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># convert the results into a dataframe</span></span>
<span><span class="va">gse.table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">gse</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>geneSetID <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><code>enrichmentScore</code> The enrichment score (ES) represents the degree to which a set S is over-represented at the top or bottom of the ranked list L.</p>
<p>We can look at the enrichment of some GO terms in GSEA plot.</p>
<div class="cell" data-hash="08_enrichment_analysis_exercise_cache/html/unnamed-chunk-11_aafbd64fee84f520473b18084793e40a">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/enrichplot/man/gseaplot2.html">gseaplot2</a></span><span class="op">(</span><span class="va">gse</span>, geneSetID <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">151</span>, <span class="fl">192</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="08_enrichment_analysis_exercise_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We have significant enriched term in the unregulated genes. Along the x axis is the ranked genes. Each of the vertical line is one of those genes that are found in this gene set. Along the running line, you can see a lot of genes running together and it falls out when the gene becomes more sparse in the list.</p>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>