<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.340">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Precourse Preparation - Theory</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./precourse_preparation_theory.html">Precourse Preparation - Theory</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Single Cell Boot Camp: Analyzing Single Cell RNA-seq with R</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./about.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Setting up</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./precourse_preparation_R.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Precourse Preparation - R</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./precourse_preparation_theory.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Precourse Preparation - Theory</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_data_import.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 1 Data Import</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 2 Quality Control</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03_normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 3 Normalization and PCA</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04_integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 4 Integration</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 5 Clustering</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06_marker_identification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 6 Marker Identification</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07_differential_expression_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 7 Differential Expression Analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08_enrichment_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 8 Enrichment Analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09_combining.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 9 Combine scRNAseq and AbSeq</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./sessioninfo.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">SessionInfo</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">1. Introduction</a></li>
  <li><a href="#bd-rhapsody-single-cell-workflow" id="toc-bd-rhapsody-single-cell-workflow" class="nav-link" data-scroll-target="#bd-rhapsody-single-cell-workflow">2. BD Rhapsody<sup>™</sup> Single Cell Workflow</a>
  <ul class="collapse">
  <li><a href="#bd-rhapsody-single-cell-system" id="toc-bd-rhapsody-single-cell-system" class="nav-link" data-scroll-target="#bd-rhapsody-single-cell-system">2.1 BD Rhapsody<sup>™</sup> Single Cell System</a></li>
  <li><a href="#overview-of-single-cell-data-analysis" id="toc-overview-of-single-cell-data-analysis" class="nav-link" data-scroll-target="#overview-of-single-cell-data-analysis">2.2 Overview of Single Cell Data Analysis</a></li>
  <li><a href="#primary-analysis---from-sequencing-data-to-expression-matrix" id="toc-primary-analysis---from-sequencing-data-to-expression-matrix" class="nav-link" data-scroll-target="#primary-analysis---from-sequencing-data-to-expression-matrix">2.3 Primary analysis - From Sequencing Data to Expression Matrix</a>
  <ul class="collapse">
  <li><a href="#bd-rhapsody-sequencing-analysis-pipeline" id="toc-bd-rhapsody-sequencing-analysis-pipeline" class="nav-link" data-scroll-target="#bd-rhapsody-sequencing-analysis-pipeline">2.3.1 BD Rhapsody<sup>™</sup> Sequencing Analysis Pipeline</a></li>
  <li><a href="#quantify-feature-expression-with-umis" id="toc-quantify-feature-expression-with-umis" class="nav-link" data-scroll-target="#quantify-feature-expression-with-umis">2.3.2 Quantify feature expression with UMIs</a></li>
  <li><a href="#output-files-from-the-primary-pipeline" id="toc-output-files-from-the-primary-pipeline" class="nav-link" data-scroll-target="#output-files-from-the-primary-pipeline">2.3.3 Output Files from the Primary Pipeline</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#additional-information-on-running-primary-pipeline" id="toc-additional-information-on-running-primary-pipeline" class="nav-link" data-scroll-target="#additional-information-on-running-primary-pipeline">3. Additional Information on Running Primary Pipeline</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Precourse Preparation - Theory</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="introduction" class="level1">
<h1>1. Introduction</h1>
<p>Our body has many cell types with different functions and they interact with each other in a daily basis. We can imagine it like a bustling city with different workers specializing in different tasks.</p>
<p>Single-cell RNA sequencing (scRNAseq) can help to understand the cell types and their functionality. It’s like taking a close-up picture of each cell and see which genes are active.</p>
<p>This technique can be applied to:</p>
<ul>
<li><p>Investigating different types of cells in a specific tissue. Example: Understanding the various cell types in the brain.</p></li>
<li><p>Discovering noval cell types or states. Example: Identifying unique cell types in the immune system.</p></li>
<li><p>Figuring out how genes change their activity as cells differentiate over time. Example: Tracking gene expression dynamics such as stem cells differentiation.</p></li>
<li><p>Finding genes that are active in certain cell types under certain conditions. Example: Finding genes that switch on and off in lung cells in asthma.</p></li>
<li><p>Examining how gene activity changes within a cell type under certain conditions. Example: Studying how nerve cells in the spinal cord change their gene activity in response to injury.</p></li>
</ul>
</section>
<section id="bd-rhapsody-single-cell-workflow" class="level1">
<h1>2. BD Rhapsody<sup>™</sup> Single Cell Workflow</h1>
<p>The BD Rhapsody<sup>™</sup> Single Cell System enables capturing and barcoding of hundreds to thousands of single cells for the analysis of genomic and proteomic information.</p>
<p>Before proceeding with Rhapsody workflow, samples of interest are processed to have <strong>live</strong> <strong>single cells in cell suspension</strong>. It is optional to sort cells to concentrate on cell type of interest.</p>
<p>The sample preparation and experimental setup are crucial and have big impacts on data quality and the following analysis. We expect live single cells in cell suspension (avoid clumps or dead cells).</p>
<p><img src="prerequisite_figures/BD_single_cell_workflow.png" class="img-fluid"></p>
<p>Once we obtain single cell suspension, the cells are captured and barcoded by the BD Rhapsody<sup>™</sup> Single-Cell Analysis System. Each cell is assigned with a unique cell barcode and mRNA molecules of a cell are tagged with that barcode. This helps to associate the RNA molecules with the corresponding cell in downstream analysis.</p>
<p>The data analysis and interpretation come as the last piece of the workflow. The upstream steps, including experimental design and sample processing, will define the analysis process: how we approach quality control, data integration or cell annotation, etc.</p>
<p>In the next chapter, you will learn which modality can be measured on the single cell level with BD Rhapsody<sup>™</sup> System.</p>
<section id="bd-rhapsody-single-cell-system" class="level2">
<h2 class="anchored" data-anchor-id="bd-rhapsody-single-cell-system">2.1 BD Rhapsody<sup>™</sup> Single Cell System</h2>
<p><img src="prerequisite_figures/BD_Rhapsody_assays_summary.png" class="img-fluid"></p>
<p>The BD Rhapsody™ Single Cell System enables simultaneous measurement of multiple parameters in thousands of individual cells. By combining available assays, researchers can generate robust and intricate datasets to unlock powerful insights.</p>
<p><strong>Gene Expression (scRNAseq):</strong> The primary purpose of scRNAseq is to measure the gene expression in individual cells. This allows to understand which genes are active or inactive and to identify different cell types based on their gene expression profiles.</p>
<p><strong>Antibody-Based Protein Analysis (AbSeq):</strong> You can use antibodies conjugated with unique barcodes to measure protein expression alongside with gene expression in individual cells.</p>
<p><strong>Sample multiplexing with antibody-based techniques (SMK):</strong> <strong>S</strong>ample <strong>m</strong>ultiplexing <strong>k</strong>it (SMK) uses generic antibody conjugated with unique barcodes to label cells from different samples. This enables pooling of multiple samples on one cartridge.</p>
<p><strong>Immune Repertoire Profiling (Full length V(D)J TCR/BCR):</strong> Full length V(D)J assay profiles the diversity of T cell and/or B cell receptor repertoires. This is valuable for studying the immune response and identifying antigen-specific cells.</p>
<p><strong>Antigen-Specific T-cell Analysis (DCode):</strong> Immudex’s DCode technology enables the profiling of T cells that recognize specific antigens such as viral epitopes or tumor-related antigens. This expands the offering of immune profiling, aiding vaccine development and immunotherapy research.</p>
<p>From gene expression profiling to antibody-based protein analysis, all assays can be combined or performed standalone to explore cellular functions and molecular dynamics in single cell resolution.</p>
</section>
<section id="overview-of-single-cell-data-analysis" class="level2">
<h2 class="anchored" data-anchor-id="overview-of-single-cell-data-analysis">2.2 Overview of Single Cell Data Analysis</h2>
<p>In general, the analysis of single cell data can be split into two parts.</p>
<p>The <strong>primary analysis</strong> transforms raw data (usually in fastq.gz format) into an expression matrix, where each row represents a gene, each column represents a cell, and each entry in the matrix contains the molecule counts of a particular feature (gene or protein) in a cell. <strong>Secondary analysis</strong> focuses on extracting meaningful biological insights from the expression matrix.</p>
<p><img src="prerequisite_figures/primary_secondary_analyses.png" class="img-fluid"></p>
</section>
<section id="primary-analysis---from-sequencing-data-to-expression-matrix" class="level2">
<h2 class="anchored" data-anchor-id="primary-analysis---from-sequencing-data-to-expression-matrix">2.3 Primary analysis - From Sequencing Data to Expression Matrix</h2>
<p>After sequencing, one will receive raw sequecning data. The raw sequencing data is usually in fastq.gz format. You may also see fq.gz.</p>
<p>The raw sequencing data is usually handled by the software, BD Rhapsody<sup>™</sup> Primary Analysis Pipeline, to enables fast and easy data processing. In this chapter, we will cover key features of the pipeline and explain key concepts necessary to understand prior working with single cell data.</p>
<section id="bd-rhapsody-sequencing-analysis-pipeline" class="level3">
<h3 class="anchored" data-anchor-id="bd-rhapsody-sequencing-analysis-pipeline">2.3.1 BD Rhapsody<sup>™</sup> Sequencing Analysis Pipeline</h3>
<p>The BD Rhapsody<sup>™</sup> Sequencing Analysis Pipeline contains a series of computational steps to process and extract meaningful information from the raw sequencing data generated by the BD Rhapsody<sup>™</sup> Sinlge Cell System.</p>
<p>The goal of the primary analysis is to transform text-based genomic data into numerical data, and we achieve this through a series of steps in the pipeline. The steps of the pipeline are listed in simplified graphics below.</p>
<p><img src="prerequisite_figures/sequencing_pipeline.png" class="img-fluid" width="1000"></p>
<p>The main steps include decoding the unique barcodes and feature molecules associated with individual cells. Oftentimes, scRNAseq relies on <strong>u</strong>nique <strong>m</strong>olecular <strong>i</strong>dentifiers (UMIs) to quantify feature expression.</p>
</section>
<section id="quantify-feature-expression-with-umis" class="level3">
<h3 class="anchored" data-anchor-id="quantify-feature-expression-with-umis">2.3.2 Quantify feature expression with UMIs</h3>
<p><strong>U</strong>nique <strong>m</strong>olecular <strong>i</strong>dentifiers (UMIs) are used to quantify featuer molecules in a cell. Here’s how it works:</p>
<ul>
<li><p><strong>Bead structure:</strong> UMIs are small, unique sequences of nucleotides that are added to the capture sequence of beads. Each bead contains tens of thousands of unique UMIs followed by the polyA capture sequence. Each UMI is specific to the molecule it attaches to and becomes its unique tag.</p>
<p><img src="prerequisite_figures/Umis_after_capture.png" class="img-fluid"></p></li>
<li><p>Thousands RNA molecule copies are created by PCR during library preparation.</p></li>
<li><p>The RNA molecule copies, including the attached UMIs, are then sequenced.</p></li>
</ul>
<p>UMIs can help to differentiate PCR copies and we can quantify correctly the feature expression. After sequencing, we count the number of unique UMIs rather than the number of reads of a transcript as the feature expression. UMIs help to:</p>
<ul>
<li><p><strong>Remove PCR copies:</strong> Because UMIs are unique to each original RNA molecule, any duplicates (multiple copies of the same RNA molecule) with the same UMI are likely the result of PCR amplification rather than a true molecule.</p></li>
<li><p><strong>Count unique molecules:</strong> By looking at the UMIs, one can count how many unique RNA molecules are originally present in the cell.</p></li>
<li><p><strong>Quantify gene expression:</strong> By counting the unique UMIs associated with each gene, one can measure the expression level of that gene in a cell. Genes with more UMIs are more active, while genes with fewer UMIs are less active.</p></li>
</ul>
<p>In summary, UMIs are used to distinguish real feature molecules from PCR copies by providing a unique identifier for each original molecule.</p>
</section>
<section id="output-files-from-the-primary-pipeline" class="level3">
<h3 class="anchored" data-anchor-id="output-files-from-the-primary-pipeline">2.3.3 Output Files from the Primary Pipeline</h3>
<p>The primary pipeline produces many output files. Below you can find a summary of output files relating to different assays.</p>
<p><img src="prerequisite_figures/output_files_cell.png" class="img-fluid"></p>
<p>The output files from the primary analysis are the input of the secondary analysis. In the course, we will work with demo data that includes WTA, AbSeq and SMK. You will get to know how to load the expression matrix (in MEX format), how to perform quality control and other steps that are commonly included in single cell analysis in Chapter 1 - 9.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>The course will ONLY cover the secondary analysis.</p>
</div>
</div>
</section>
</section>
</section>
<section id="additional-information-on-running-primary-pipeline" class="level1">
<h1>3. Additional Information on Running Primary Pipeline</h1>
<p>Here is how to set up and run the BD Rhapsody™ Analysis pipelines: BD® Single-Cell Multiomics Analysis Setup User Guide.&nbsp; (<a href="https://scomix.bd.com/hc/article_attachments/16954103767949" class="uri">https://scomix.bd.com/hc/article_attachments/16954103767949</a>)<br>
<br>
- This guide provides detailed instructions on how to set up and run the BD Rhapsody™ Analysis pipelines on the Seven Bridges Genomics platform or a local installation.<br>
&nbsp;<br>
Here is the bioinformatics handbook as a resource: BD® Single-Cell Multiomics Bioinformatics Handbook. (<a href="https://scomix.bd.com/hc/article_attachments/16954089124621" class="uri">https://scomix.bd.com/hc/article_attachments/16954089124621</a>)<br>
<br>
- This handbook is a comprehensive reference to help you prepare and analyze single-cell libraries with the BD Rhapsody™ Single-Cell Analysis system or the BD Rhapsody™ Express Single-Cell Analysis system. Major aspects of the BD® Single-Cell Multiomics bioinformatics workflow are covered. This reference explains the BD single-cell multiomics sequencing analysis algorithms to deepen your understanding of how single-cell mRNA and protein (AbSeq) expression profiles are generated. In addition, the handbook defines every analysis metric.<br>
</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>