<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.340">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Precourse Preparation - R</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script><link rel="stylesheet" href="styles.css">
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./precourse_preparation_R.html">Precourse Preparation - R</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Single Cell Boot Camp: Analyzing Single Cell RNA-seq with R</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./about.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Setting up</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./precourse_preparation_R.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Precourse Preparation - R</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./precourse_preparation_theory.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Precourse Preparation - Theory</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_data_import.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 1 Data Import</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./02_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 2 Quality Control</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./03_normalization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 3 Normalization and PCA</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04_integration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 4 Integration</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./05_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 5 Clustering</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06_marker_identification.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 6 Marker Identification</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./07_differential_expression_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 7 Differential Expression Analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./08_enrichment_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 8 Enrichment Analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./09_combining.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Chapter 9 Combine scRNAseq and AbSeq</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./sessioninfo.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">SessionInfo</span></a>
  </div>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li>
<a href="#data-frame" id="toc-data-frame" class="nav-link" data-scroll-target="#data-frame">1. Data frame</a>
  <ul class="collapse">
<li><a href="#what-is-data-frame" id="toc-what-is-data-frame" class="nav-link" data-scroll-target="#what-is-data-frame">1.1 What is data frame</a></li>
  <li><a href="#convenient-version-of-data-frame-tibble" id="toc-convenient-version-of-data-frame-tibble" class="nav-link" data-scroll-target="#convenient-version-of-data-frame-tibble">1.2 Convenient version of data frame: “Tibble”</a></li>
  <li>
<a href="#working-with-data-frames-in-r" id="toc-working-with-data-frames-in-r" class="nav-link" data-scroll-target="#working-with-data-frames-in-r">1.3 Working with data frames in R</a>
  <ul class="collapse">
<li><a href="#example-data" id="toc-example-data" class="nav-link" data-scroll-target="#example-data">1.3.1 Example data</a></li>
  <li><a href="#loading-data" id="toc-loading-data" class="nav-link" data-scroll-target="#loading-data">1.3.2 Loading data</a></li>
  <li><a href="#viewing-the-content" id="toc-viewing-the-content" class="nav-link" data-scroll-target="#viewing-the-content">1.3.3 Viewing the content</a></li>
  <li><a href="#extracting-information" id="toc-extracting-information" class="nav-link" data-scroll-target="#extracting-information">1.3.4 Extracting information</a></li>
  <li><a href="#modifying-information" id="toc-modifying-information" class="nav-link" data-scroll-target="#modifying-information">1.3.4 Modifying information</a></li>
  </ul>
</li>
  </ul>
</li>
  <li>
<a href="#tidyverse" id="toc-tidyverse" class="nav-link" data-scroll-target="#tidyverse">2. Tidyverse</a>
  <ul class="collapse">
<li><a href="#pipe" id="toc-pipe" class="nav-link" data-scroll-target="#pipe">2.1 Pipe</a></li>
  <li><a href="#example-data-1" id="toc-example-data-1" class="nav-link" data-scroll-target="#example-data-1">2.2 Example data</a></li>
  <li><a href="#select" id="toc-select" class="nav-link" data-scroll-target="#select">2.3 <code>select()</code></a></li>
  <li><a href="#mutate" id="toc-mutate" class="nav-link" data-scroll-target="#mutate">2.4 <code>mutate()</code></a></li>
  <li><a href="#summarise" id="toc-summarise" class="nav-link" data-scroll-target="#summarise">2.5 <code>summarise()</code></a></li>
  <li><a href="#if_else" id="toc-if_else" class="nav-link" data-scroll-target="#if_else">2.6 <code>if_else()</code></a></li>
  <li><a href="#count" id="toc-count" class="nav-link" data-scroll-target="#count">2.7 <code>count()</code></a></li>
  <li><a href="#group_by" id="toc-group_by" class="nav-link" data-scroll-target="#group_by">2.8 <code>group_by()</code></a></li>
  <li><a href="#filter" id="toc-filter" class="nav-link" data-scroll-target="#filter">2.8 <code>filter()</code></a></li>
  <li><a href="#rename" id="toc-rename" class="nav-link" data-scroll-target="#rename">2.9 <code>rename()</code></a></li>
  </ul>
</li>
  <li>
<a href="#seurat-object" id="toc-seurat-object" class="nav-link" data-scroll-target="#seurat-object">3. Seurat object</a>
  <ul class="collapse">
<li><a href="#seurat-object-structure" id="toc-seurat-object-structure" class="nav-link" data-scroll-target="#seurat-object-structure">3.1 Seurat Object Structure</a></li>
  <li><a href="#assays" id="toc-assays" class="nav-link" data-scroll-target="#assays">3.2 Assays</a></li>
  <li><a href="#meta-data" id="toc-meta-data" class="nav-link" data-scroll-target="#meta-data">3.3 Meta Data</a></li>
  </ul>
</li>
  <li><a href="#references-and-resources" id="toc-references-and-resources" class="nav-link" data-scroll-target="#references-and-resources">4. References and Resources</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">Precourse Preparation - R</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-1_ea13b0cabcb175a36836f87ed21bf084">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># libraries for this chapter</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readxl.tidyverse.org">readxl</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="introduction" class="level1"><h1>Introduction</h1>
<p>This tutorial will help you to understand the basic concepts in R. It will cover the following topics:</p>
<ol type="1">
<li>Working with Dataframes</li>
<li>Tidyverse library</li>
<li>Understanding the Seurat object</li>
</ol></section><section id="data-frame" class="level1"><h1>1. Data frame</h1>
<section id="what-is-data-frame" class="level2"><h2 class="anchored" data-anchor-id="what-is-data-frame">1.1 What is data frame</h2>
<p>Data frame is data structure in R that store values of any type. You can think of it as a table with columns and rows storing the data as in below example.</p>
<p><img src="prerequisite_figures/data_frame.png" class="img-fluid"></p>
<p>Credit: https://www.geeksforgeeks.org/r-data-frames/</p>
<p>Often in data science language the columns are named as variables, the rows are observations and each entry in a table is a value.</p>
<p>Some of the most common examples of input files that can be loaded into R as data frames are spreadsheets, TSVs, CSVs files. TSV and CSV stands for tab-separated or coma-separated values.</p>
<p>Another way to view data frame is as list of vectors of the same length. The vector can be defined as type of data structure that contains elements of the same type. Below you can find example on how create the table presented in Figure 1 from vectors.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-2_3b71934de04094eaf789ad5d16d8ad81">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#define vectors which will create columns with data in data frame</span></span>
<span></span>
<span><span class="co">## Vector of names</span></span>
<span><span class="va">Name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Avrey Bradley"</span>,<span class="st">"John Holland"</span>,<span class="st">"Jonas Jerebko"</span>,<span class="st">"Jordan Midkey"</span>,<span class="st">"Terry Rozier"</span>,<span class="st">"Jared Sullinger"</span>,<span class="st">"Evan Turner"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Vector of team's names</span></span>
<span><span class="va">Team</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Boston Celtics"</span>, <span class="st">"Boston Celtics"</span>,<span class="st">"Boston Celtics"</span>,<span class="st">"Boston Celtics"</span>,<span class="st">"Boston Celtics"</span>,<span class="st">"Boston Celtics"</span>,<span class="st">"Boston Celtics"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Vector of values</span></span>
<span><span class="va">Number</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.0</span>,<span class="fl">30.0</span>,<span class="fl">8.0</span>,<span class="cn">NaN</span>,<span class="fl">12.0</span>,<span class="fl">7.0</span>,<span class="fl">11.0</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Vector of positions</span></span>
<span><span class="va">Position</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"PG"</span>,<span class="st">"SG"</span>,<span class="st">"PF"</span>,<span class="st">"PF"</span>,<span class="st">"PG"</span>,<span class="st">"C"</span>,<span class="st">"SG"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Vector of age</span></span>
<span><span class="va">Age</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">25.0</span>,<span class="fl">27.0</span>,<span class="fl">29.0</span>,<span class="fl">21.0</span>,<span class="fl">22.0</span>,<span class="cn">NaN</span>,<span class="fl">27.0</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#create data frame from vectors</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">Name</span>,<span class="va">Team</span>,<span class="va">Number</span>,<span class="va">Position</span>,<span class="va">Age</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             Name           Team Number Position Age
1   Avrey Bradley Boston Celtics      0       PG  25
2    John Holland Boston Celtics     30       SG  27
3   Jonas Jerebko Boston Celtics      8       PF  29
4   Jordan Midkey Boston Celtics    NaN       PF  21
5    Terry Rozier Boston Celtics     12       PG  22
6 Jared Sullinger Boston Celtics      7        C NaN
7     Evan Turner Boston Celtics     11       SG  27</code></pre>
</div>
</div>
</section><section id="convenient-version-of-data-frame-tibble" class="level2"><h2 class="anchored" data-anchor-id="convenient-version-of-data-frame-tibble">1.2 Convenient version of data frame: “Tibble”</h2>
<p>The tibble is package that manipulate data frames. The tibble type of table looks like data frame but have some advantages over using regular data frames.</p>
<p>For example, while loading data as classical data frame, the spaces in names are conversed to periods or “x” before numeric columns is added. While using tibbles, the input names and types remain unchanged. In this tutorial we will work mainly with tibble data frame type.</p>
</section><section id="working-with-data-frames-in-r" class="level2"><h2 class="anchored" data-anchor-id="working-with-data-frames-in-r">1.3 Working with data frames in R</h2>
<section id="example-data" class="level3"><h3 class="anchored" data-anchor-id="example-data">1.3.1 Example data</h3>
<p>In this part of tutorial, we will work on table downloaded from supplementary data of Salcher, Stefan et al.&nbsp;publication (doi:10.1016/j.ccell.2022.10.008). The table contains the metadata of patients for which samples were processed on different single-cell platforms.</p>
</section><section id="loading-data" class="level3"><h3 class="anchored" data-anchor-id="loading-data">1.3.2 Loading data</h3>
<p>Depending on the input format, there are different dedicated libraries to load data into R. Here we start from excel spreadsheet and use function from <code>readxl</code> library. The <code>readxl</code> library is part of tidyverse packages that will be covered in second chapter of this tutorial.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-3_e1dda88f81f0b8a9a5670780abbeb2f8">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html">read_excel</a></span><span class="op">(</span><span class="st">"data/precourse_preparation_data/mmc4.xlsx"</span>, skip <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>NOTE:</strong></p>
<ol type="1">
<li><p>The two first rows in the table are empty and therefore skipped while loading.</p></li>
<li><p>One of the columns in excel spreadsheet is empty (column N). The R assign automatically the default name to that column and keep it within data frame.</p></li>
<li><p>All the missing entries in rows will be filled with “NA” values automatically.</p></li>
</ol></section><section id="viewing-the-content" class="level3"><h3 class="anchored" data-anchor-id="viewing-the-content">1.3.3 Viewing the content</h3>
<p>There are different ways to view the content of data frame. You can directly click on <code>df_example</code> to display the content.</p>
<p>The “str” function lists directly in the console: the type of data (tibble), dimensions (322x18), columns with first values displayed and type of values.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-4_7e0444c22b1e352579347d90ce755daa">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">df_example</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>tibble [322 × 18] (S3: tbl_df/tbl/data.frame)
 $ ...1                    : num [1:322] 0 1 2 3 4 5 6 7 8 9 ...
 $ study                   : chr [1:322] "Adams_Kaminski_2020" "Adams_Kaminski_2020" "Adams_Kaminski_2020" "Adams_Kaminski_2020" ...
 $ dataset                 : chr [1:322] "Adams_Kaminski_2020" "Adams_Kaminski_2020" "Adams_Kaminski_2020" "Adams_Kaminski_2020" ...
 $ patient                 : chr [1:322] "Adams_Kaminski_2020_001C" "Adams_Kaminski_2020_002C" "Adams_Kaminski_2020_003C" "Adams_Kaminski_2020_8CO" ...
 $ uicc_stage              : chr [1:322] NA NA NA NA ...
 $ tumor_stage             : chr [1:322] NA NA NA NA ...
 $ sex                     : chr [1:322] "male" "female" "female" "male" ...
 $ ever_smoker             : chr [1:322] "no" "no" "no" "yes" ...
 $ driver_genes            : chr [1:322] NA NA NA NA ...
 $ condition               : chr [1:322] "non-cancer" "non-cancer" "non-cancer" "COPD" ...
 $ age                     : num [1:322] 22 25 67 65 66 49 62 57 66 20 ...
 $ platform                : chr [1:322] "10x" "10x" "10x" "10x" ...
 $ platform_fine           : chr [1:322] "10x_3p_v2" "10x_3p_v2" "10x_3p_v2" "10x_3p_v2" ...
 $ ...14                   : logi [1:322] NA NA NA NA NA NA ...
 $ tissue: normal          : num [1:322] 1 1 1 1 1 1 1 1 1 1 ...
 $ tissue: normal_adjacent : num [1:322] 0 0 0 0 0 0 0 0 0 0 ...
 $ tissue: tumor_primary   : num [1:322] 0 0 0 0 0 0 0 0 0 0 ...
 $ tissue: tumor_metastasis: num [1:322] 0 0 0 0 0 0 0 0 0 0 ...</code></pre>
</div>
</div>
</section><section id="extracting-information" class="level3"><h3 class="anchored" data-anchor-id="extracting-information">1.3.4 Extracting information</h3>
<p>Extracting data from an R data frame means to access its rows or columns. First let’s list all columns available in the dataset.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-5_f6ee30ec309c278be3760bc8ae893b26">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">df_example</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "...1"                     "study"                   
 [3] "dataset"                  "patient"                 
 [5] "uicc_stage"               "tumor_stage"             
 [7] "sex"                      "ever_smoker"             
 [9] "driver_genes"             "condition"               
[11] "age"                      "platform"                
[13] "platform_fine"            "...14"                   
[15] "tissue: normal"           "tissue: normal_adjacent" 
[17] "tissue: tumor_primary"    "tissue: tumor_metastasis"</code></pre>
</div>
</div>
<p>To extract values from single column of the data frame different options can be used:</p>
<ul>
<li>Using $ operator and providing the column name</li>
</ul>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-6_1212d34eaab5759a2d367ed979b14339">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">df_example</span><span class="op">$</span><span class="va">study</span><span class="op">)</span> <span class="co"># head allows to display only 10 first elements </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Adams_Kaminski_2020" "Adams_Kaminski_2020" "Adams_Kaminski_2020"
[4] "Adams_Kaminski_2020" "Adams_Kaminski_2020" "Adams_Kaminski_2020"</code></pre>
</div>
</div>
<ul>
<li>Using [ , ] and providing the column number (index)</li>
</ul>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-7_57fb84db338a3224fd19d3f7a0569517">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df_example</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># head allows to display only 10 first elements </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 1
  study              
  &lt;chr&gt;              
1 Adams_Kaminski_2020
2 Adams_Kaminski_2020
3 Adams_Kaminski_2020
4 Adams_Kaminski_2020
5 Adams_Kaminski_2020
6 Adams_Kaminski_2020</code></pre>
</div>
</div>
<ul>
<li>Using [[]] and providing column number (index)</li>
</ul>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-8_327bca4802e8bd4651dc38dd05cc36f9">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df_example</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Adams_Kaminski_2020" "Adams_Kaminski_2020" "Adams_Kaminski_2020"
[4] "Adams_Kaminski_2020" "Adams_Kaminski_2020" "Adams_Kaminski_2020"</code></pre>
</div>
</div>
<ul>
<li>Using column name</li>
</ul>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-9_d7da9f0323ae2dccbec24c62f684c327">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df_example</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"study"</span><span class="op">)</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 1
  study              
  &lt;chr&gt;              
1 Adams_Kaminski_2020
2 Adams_Kaminski_2020
3 Adams_Kaminski_2020
4 Adams_Kaminski_2020
5 Adams_Kaminski_2020
6 Adams_Kaminski_2020</code></pre>
</div>
</div>
<p>To extract multiple columns or rows at the same time the following rules apply:</p>
<p><img src="prerequisite_figures/df_rows_columns.jpg" class="img-fluid"></p>
<ul>
<li>Extract rows range, include all columns.</li>
</ul>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-10_153d79279461f78d69f822171ef82676">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df_example</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span>,<span class="op">]</span> <span class="co">#display rows from 2-4 and all columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 18
   ...1 study           dataset patient uicc_stage tumor_stage sex   ever_smoker
  &lt;dbl&gt; &lt;chr&gt;           &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;      
1     1 Adams_Kaminski… Adams_… Adams_… &lt;NA&gt;       &lt;NA&gt;        fema… no         
2     2 Adams_Kaminski… Adams_… Adams_… &lt;NA&gt;       &lt;NA&gt;        fema… no         
3     3 Adams_Kaminski… Adams_… Adams_… &lt;NA&gt;       &lt;NA&gt;        male  yes        
# ℹ 10 more variables: driver_genes &lt;chr&gt;, condition &lt;chr&gt;, age &lt;dbl&gt;,
#   platform &lt;chr&gt;, platform_fine &lt;chr&gt;, ...14 &lt;lgl&gt;, `tissue: normal` &lt;dbl&gt;,
#   `tissue: normal_adjacent` &lt;dbl&gt;, `tissue: tumor_primary` &lt;dbl&gt;,
#   `tissue: tumor_metastasis` &lt;dbl&gt;</code></pre>
</div>
</div>
<ul>
<li>Extract rows, include all columns.</li>
</ul>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-11_6cc2067a1db622e0c3786e185c673644">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df_example</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span><span class="op">)</span>,<span class="op">]</span> <span class="co">#display rows 2 and 4 and all columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 18
   ...1 study           dataset patient uicc_stage tumor_stage sex   ever_smoker
  &lt;dbl&gt; &lt;chr&gt;           &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;      
1     1 Adams_Kaminski… Adams_… Adams_… &lt;NA&gt;       &lt;NA&gt;        fema… no         
2     3 Adams_Kaminski… Adams_… Adams_… &lt;NA&gt;       &lt;NA&gt;        male  yes        
# ℹ 10 more variables: driver_genes &lt;chr&gt;, condition &lt;chr&gt;, age &lt;dbl&gt;,
#   platform &lt;chr&gt;, platform_fine &lt;chr&gt;, ...14 &lt;lgl&gt;, `tissue: normal` &lt;dbl&gt;,
#   `tissue: normal_adjacent` &lt;dbl&gt;, `tissue: tumor_primary` &lt;dbl&gt;,
#   `tissue: tumor_metastasis` &lt;dbl&gt;</code></pre>
</div>
</div>
<ul>
<li>Extract columns range, include all rows.</li>
</ul>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-12_69e94a405c65ba0aebba76863deb578a">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df_example</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span> <span class="co">#display columns from 2-4 and all rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  study               dataset             patient                 
  &lt;chr&gt;               &lt;chr&gt;               &lt;chr&gt;                   
1 Adams_Kaminski_2020 Adams_Kaminski_2020 Adams_Kaminski_2020_001C
2 Adams_Kaminski_2020 Adams_Kaminski_2020 Adams_Kaminski_2020_002C
3 Adams_Kaminski_2020 Adams_Kaminski_2020 Adams_Kaminski_2020_003C
4 Adams_Kaminski_2020 Adams_Kaminski_2020 Adams_Kaminski_2020_8CO 
5 Adams_Kaminski_2020 Adams_Kaminski_2020 Adams_Kaminski_2020_23CO
6 Adams_Kaminski_2020 Adams_Kaminski_2020 Adams_Kaminski_2020_034C</code></pre>
</div>
</div>
<ul>
<li>Extract columns, include all rows.</li>
</ul>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-13_d5c741d5492191f11da2aaa9abb863aa">
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df_example</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span><span class="op">)</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span> <span class="co">#display columns 2 and 4 and all rows</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 2
  study               patient                 
  &lt;chr&gt;               &lt;chr&gt;                   
1 Adams_Kaminski_2020 Adams_Kaminski_2020_001C
2 Adams_Kaminski_2020 Adams_Kaminski_2020_002C
3 Adams_Kaminski_2020 Adams_Kaminski_2020_003C
4 Adams_Kaminski_2020 Adams_Kaminski_2020_8CO 
5 Adams_Kaminski_2020 Adams_Kaminski_2020_23CO
6 Adams_Kaminski_2020 Adams_Kaminski_2020_034C</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><strong>Exercise</strong></li>
</ul>
<p>Extract <strong>rows</strong> from 5-10 and <strong>columns</strong>&nbsp;from 3-8.</p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<ul>
<li><strong>Answer</strong></li>
</ul>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-14_94a4237d59453668089ce641d7df536a">
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df_example</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">8</span><span class="op">)</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section><section id="modifying-information" class="level3"><h3 class="anchored" data-anchor-id="modifying-information">1.3.4 Modifying information</h3>
<p>The columns or rows can be removed or added to data frame as needed. It is possible to drop specific columns indicating the column indexes.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-15_84ec7f9f3511d45aa860b529c17f9cd0">
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">modified_df_example</span> <span class="op">&lt;-</span> <span class="va">df_example</span><span class="op">[</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">5</span><span class="op">)</span><span class="op">]</span>    <span class="co"># Drop columns 1,2,5 and include all columns</span></span>
<span><span class="va">modified_df_example</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span> <span class="co"># data frame with excluded columns </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 15
  dataset     patient tumor_stage sex   ever_smoker driver_genes condition   age
  &lt;chr&gt;       &lt;chr&gt;   &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;       &lt;chr&gt;        &lt;chr&gt;     &lt;dbl&gt;
1 Adams_Kami… Adams_… &lt;NA&gt;        male  no          &lt;NA&gt;         non-canc…    22
2 Adams_Kami… Adams_… &lt;NA&gt;        fema… no          &lt;NA&gt;         non-canc…    25
3 Adams_Kami… Adams_… &lt;NA&gt;        fema… no          &lt;NA&gt;         non-canc…    67
4 Adams_Kami… Adams_… &lt;NA&gt;        male  yes         &lt;NA&gt;         COPD         65
5 Adams_Kami… Adams_… &lt;NA&gt;        male  yes         &lt;NA&gt;         COPD         66
6 Adams_Kami… Adams_… &lt;NA&gt;        male  yes         &lt;NA&gt;         non-canc…    49
# ℹ 7 more variables: platform &lt;chr&gt;, platform_fine &lt;chr&gt;, ...14 &lt;lgl&gt;,
#   `tissue: normal` &lt;dbl&gt;, `tissue: normal_adjacent` &lt;dbl&gt;,
#   `tissue: tumor_primary` &lt;dbl&gt;, `tissue: tumor_metastasis` &lt;dbl&gt;</code></pre>
</div>
</div>
<p>The same principle apply to excluding the rows</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-16_40d894f216f81f0e3bc169e19e09b571">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">modified_df_example</span> <span class="op">&lt;-</span> <span class="va">df_example</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">5</span><span class="op">)</span>,<span class="op">]</span>     <span class="co"># Drop rows 1,2,5 and include all columns</span></span>
<span><span class="va">modified_df_example</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 18
   ...1 study           dataset patient uicc_stage tumor_stage sex   ever_smoker
  &lt;dbl&gt; &lt;chr&gt;           &lt;chr&gt;   &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;      
1     2 Adams_Kaminski… Adams_… Adams_… &lt;NA&gt;       &lt;NA&gt;        fema… no         
2     3 Adams_Kaminski… Adams_… Adams_… &lt;NA&gt;       &lt;NA&gt;        male  yes        
3     5 Adams_Kaminski… Adams_… Adams_… &lt;NA&gt;       &lt;NA&gt;        male  yes        
4     6 Adams_Kaminski… Adams_… Adams_… &lt;NA&gt;       &lt;NA&gt;        fema… yes        
5     7 Adams_Kaminski… Adams_… Adams_… &lt;NA&gt;       &lt;NA&gt;        fema… yes        
6     8 Adams_Kaminski… Adams_… Adams_… &lt;NA&gt;       &lt;NA&gt;        fema… no         
# ℹ 10 more variables: driver_genes &lt;chr&gt;, condition &lt;chr&gt;, age &lt;dbl&gt;,
#   platform &lt;chr&gt;, platform_fine &lt;chr&gt;, ...14 &lt;lgl&gt;, `tissue: normal` &lt;dbl&gt;,
#   `tissue: normal_adjacent` &lt;dbl&gt;, `tissue: tumor_primary` &lt;dbl&gt;,
#   `tissue: tumor_metastasis` &lt;dbl&gt;</code></pre>
</div>
</div>
<p>We can remove column with specific name. To drop column, we chose all the columns (called by “colnames”) except (!=) “patient”.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-17_2562beb0ef53440ee5ab58722ffaca0a">
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">modified_df_example</span> <span class="op">&lt;-</span> <span class="va">df_example</span><span class="op">[</span> , <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">df_example</span><span class="op">)</span> <span class="op">!=</span> <span class="st">"patient"</span><span class="op">]</span>    <span class="co"># Drop column "patient"</span></span>
<span><span class="va">modified_df_example</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 17
   ...1 study      dataset uicc_stage tumor_stage sex   ever_smoker driver_genes
  &lt;dbl&gt; &lt;chr&gt;      &lt;chr&gt;   &lt;chr&gt;      &lt;chr&gt;       &lt;chr&gt; &lt;chr&gt;       &lt;chr&gt;       
1     0 Adams_Kam… Adams_… &lt;NA&gt;       &lt;NA&gt;        male  no          &lt;NA&gt;        
2     1 Adams_Kam… Adams_… &lt;NA&gt;       &lt;NA&gt;        fema… no          &lt;NA&gt;        
3     2 Adams_Kam… Adams_… &lt;NA&gt;       &lt;NA&gt;        fema… no          &lt;NA&gt;        
4     3 Adams_Kam… Adams_… &lt;NA&gt;       &lt;NA&gt;        male  yes         &lt;NA&gt;        
5     4 Adams_Kam… Adams_… &lt;NA&gt;       &lt;NA&gt;        male  yes         &lt;NA&gt;        
6     5 Adams_Kam… Adams_… &lt;NA&gt;       &lt;NA&gt;        male  yes         &lt;NA&gt;        
# ℹ 9 more variables: condition &lt;chr&gt;, age &lt;dbl&gt;, platform &lt;chr&gt;,
#   platform_fine &lt;chr&gt;, ...14 &lt;lgl&gt;, `tissue: normal` &lt;dbl&gt;,
#   `tissue: normal_adjacent` &lt;dbl&gt;, `tissue: tumor_primary` &lt;dbl&gt;,
#   `tissue: tumor_metastasis` &lt;dbl&gt;</code></pre>
</div>
</div>
<p>The new columns can be added to data frame. To add new column, use $ to indicate the column name for a new column as in the example below.</p>
<p>Here we assign the values from already existing column to a new column. The new column is added at the end of the data table.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-18_2295df84157ef87ee5bd38278eba3f2f">
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df_example</span><span class="op">$</span><span class="va">technology_type</span> <span class="op">&lt;-</span> <span class="va">df_example</span><span class="op">$</span><span class="va">platform</span> <span class="co"># add new column named "technology_type" with values from "platform" existing column</span></span>
<span><span class="va">df_example</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">df_example</span><span class="op">)</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span> <span class="co">#show all rows and last column</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 1
  technology_type
  &lt;chr&gt;          
1 10x            
2 10x            
3 10x            
4 10x            
5 10x            
6 10x            </code></pre>
</div>
</div>
<p>The values from chosen column can be replaced with other values. In example below we search in df_example data frame in technology_type column for rows that contain “BD-Rhapsody” text. For those values we assign “microwell-based”.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-19_204c1f41d8747927ae345fa2da5c4bb8">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df_example</span><span class="op">$</span><span class="va">technology_type</span><span class="op">[</span><span class="va">df_example</span><span class="op">$</span><span class="va">technology_type</span> <span class="op">==</span> <span class="st">"BD-Rhapsody"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"microwell-based"</span></span>
<span><span class="va">df_example</span><span class="op">$</span><span class="va">technology_type</span><span class="op">[</span><span class="va">df_example</span><span class="op">$</span><span class="va">technology_type</span> <span class="op">==</span> <span class="st">"10x"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"droplet-based"</span></span>
<span></span>
<span><span class="va">df_example</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"platform"</span>,<span class="st">"technology_type"</span><span class="op">)</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span> <span class="co">#display columns</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 2
  platform technology_type
  &lt;chr&gt;    &lt;chr&gt;          
1 10x      droplet-based  
2 10x      droplet-based  
3 10x      droplet-based  
4 10x      droplet-based  
5 10x      droplet-based  
6 10x      droplet-based  </code></pre>
</div>
</div>
<p>The last useful function to cover in this chapter is summary(). The function gives the statistical summary of each of the columns when applied on data frame. It can be applied on slice of data as for example only on one column or chosen columns and rows.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-20_b8677d62fc53457a93e7a9b745ec6062">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">df_example</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      ...1           study             dataset            patient         
 Min.   :  0.00   Length:322         Length:322         Length:322        
 1st Qu.: 80.25   Class :character   Class :character   Class :character  
 Median :160.50   Mode  :character   Mode  :character   Mode  :character  
 Mean   :160.50                                                           
 3rd Qu.:240.75                                                           
 Max.   :321.00                                                           
                                                                          
  uicc_stage        tumor_stage            sex            ever_smoker       
 Length:322         Length:322         Length:322         Length:322        
 Class :character   Class :character   Class :character   Class :character  
 Mode  :character   Mode  :character   Mode  :character   Mode  :character  
                                                                            
                                                                            
                                                                            
                                                                            
 driver_genes        condition              age          platform        
 Length:322         Length:322         Min.   :20.00   Length:322        
 Class :character   Class :character   1st Qu.:56.00   Class :character  
 Mode  :character   Mode  :character   Median :63.00   Mode  :character  
                                       Mean   :60.73                     
                                       3rd Qu.:69.00                     
                                       Max.   :90.00                     
                                       NA's   :89                        
 platform_fine       ...14         tissue: normal   tissue: normal_adjacent
 Length:322         Mode:logical   Min.   :0.0000   Min.   :0.0000         
 Class :character   NA's:322       1st Qu.:0.0000   1st Qu.:0.0000         
 Mode  :character                  Median :0.0000   Median :0.0000         
                                   Mean   :0.2267   Mean   :0.3261         
                                   3rd Qu.:0.0000   3rd Qu.:1.0000         
                                   Max.   :1.0000   Max.   :1.0000         
                                                                           
 tissue: tumor_primary tissue: tumor_metastasis technology_type   
 Min.   :0.0000        Min.   :0.0000           Length:322        
 1st Qu.:0.0000        1st Qu.:0.0000           Class :character  
 Median :1.0000        Median :0.0000           Mode  :character  
 Mean   :0.5497        Mean   :0.1398                             
 3rd Qu.:1.0000        3rd Qu.:0.0000                             
 Max.   :1.0000        Max.   :1.0000                             
                                                                  </code></pre>
</div>
</div>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-21_c23d08c9aa3334823087321ce07dc1d7">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">df_example</span><span class="op">$</span><span class="va">age</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
  20.00   56.00   63.00   60.73   69.00   90.00      89 </code></pre>
</div>
</div>
</section></section></section><section id="tidyverse" class="level1"><h1>2. Tidyverse</h1>
<p>Data often need to be re-shaped, re-arranged, and re-formatted, so that they can be visualized or used in downstream analysis. Throughout this chapter you will learn how to use the tidyverse library to navigate through and adjust the tables you work with.</p>
<p>The tidyverse is collection of R packages. The core packages included in tidyverse are ggplot2, tidyr, readr, dplyr, stringr, purrr, and forcats. Each of the packages is dedicated to handle different operations. In below graphics you can find a summary of all packages and their core functionalities.</p>
<p><img src="prerequisite_figures/tidyverse_libs_summary_graphics.png" class="img-fluid"></p>
<p>Credit: https://www.storybench.org/getting-started-with-tidyverse-in-r/</p>
<p>In this chapter we will cover some of the functions included in tidyverse library and used during workshop. If you would like to learn more about tidyverse and additional libraries, we encourage you to visit official website <a href="https://www.tidyverse.org/" class="uri">https://www.tidyverse.org/</a>.</p>
<p>When tidyverse is loaded you will see the list of packages included and loaded with tidyverse.</p>
<p>After loading tidyverse, the R inform you about conflicting packages. The reason is there are other packages already loaded by default in R.</p>
<p>Packages consist of a collection of functions related to a particular purpose. Those functions can have naming conflicts with functions from other packages.</p>
<p>The two functions from stats package: filer and lag, exist as well in tidyverse. After loading tidyverse package, the functions from stats package are masked and by default the tidyverse functions will be used. If you would like to use filter() or lag() functions from stats package, you should indicate the package name as in example below.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-22_417f4f81aa794f251ae508d4c69b03e4">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Function in this example are empty for illustration purposes. To use that function you need to provide arguments inside.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Without indicating the package name in front of function name, R will use the package that was loaded the most recently in your R session.</p>
<section id="pipe" class="level2"><h2 class="anchored" data-anchor-id="pipe">2.1 Pipe</h2>
<p>A pipe is a type of operator in R <code>%&gt;%</code>. It takes the output of one function and passes it as the first argument of the next function, allowing us to chain together several steps in R.</p>
<p><img src="prerequisite_figures/pipe_tidyverse.png" class="img-fluid"></p>
<p>Credit: tidyverse.org</p>
<p>The pipe sends the data on left side of the pipe to be the first argument of the function on the right side of the pipe.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-23_cd0242fa0d7d4e273e8f018fc005ae7e">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fl">16</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># is equivalent to sqrt(16)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4</code></pre>
</div>
</div>
<p>We can pipe multiple functions together. The operations are performed from the left to the right sequentially.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-24_d2947c348157532b5001ca023e40ad1a">
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fl">16</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># is equivalent to log2(sqrt(16))</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
</div>
<p>Below you can find exercises with more examples of piping mathematical functions in R together.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>Exercise: What would be the equivalent mathematical expression to below piped sequence of operations?</li>
</ol>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-25_bcb384f238fbe106f340655dce96227f">
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fl">26</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="op">-</span><span class="fl">22</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="op">+</span><span class="fl">4</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<ol type="1">
<li>Answer:</li>
</ol>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-26_f82b81ad2328022b14b32e85adb9ea4b">
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="fl">26</span><span class="op">-</span><span class="fl">22</span><span class="op">)</span> <span class="op">+</span> <span class="fl">4</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<ol start="2" type="1">
<li>Exercise: How below mathematical operation can be shown with pipe option?</li>
</ol>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-27_96ec07c87b69ccbe7b07bc591b7fe5a6">
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="fl">64</span><span class="op">)</span><span class="op">)</span> <span class="op">-</span> <span class="fl">8</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<ol start="2" type="1">
<li>Answer</li>
</ol>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-28_e50ba5610b70edb78f4ad4a9dd86c73b">
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fl">64</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="op">-</span><span class="fl">8</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section><section id="example-data-1" class="level2"><h2 class="anchored" data-anchor-id="example-data-1">2.2 Example data</h2>
<p>In this chapter we will work on an expression matrix table which is in CSV format. Below is example of how to load the table.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-29_a0da19e499a0a98daabbfb6e0adc9065">
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"data/precourse_preparation_data/Demo-WTA-AbSeq.csv"</span>, sep <span class="op">=</span> <span class="st">","</span>, header<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># file: provide full directory to file</span></span>
<span><span class="co"># sep: indicate separator, in this case coma</span></span>
<span><span class="co"># header = True is indicating to take the first row as a column names</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Displaying first rows and columns of the table gives us a glimpse on what the data contain.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-30_2df643b5cd8ac0580059884a9393a7ae">
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">expMat</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         V1 A1BG A1BG.AS1 A1CF A2M
1:  3261781    0        0    0   0
2:  9155806    0        0    0   0
3:  4304661    0        0    0   0
4:  2216902    0        0    0   0
5: 13160139    0        0    0   0</code></pre>
</div>
</div>
<p>The gene names are displayed in columns and rows contain information about the cells. The column named V1 hold information about the cell index that is given to each cell in primary analysis.</p>
<p>We will set the V1 column as row names. In that way all the columns will contain the gene or protein counts.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-31_15d0199e1463b351d2ad4272f01ef8b2">
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#use first column as row names</span></span>
<span><span class="va">expMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">expMat</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The same set of above operations can be performed with pipe in one line of code. We read the file and convert first column to row names and save as data frame. The result is assigned to expMat variable by using the arrow.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-32_8c0ceff125d0a3442416500b5fcf7431">
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#pipe operations</span></span>
<span><span class="va">expMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span>file <span class="op">=</span> <span class="st">"data/precourse_preparation_data/Demo-WTA-AbSeq.csv"</span>, sep <span class="op">=</span> <span class="st">","</span>, header<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="select" class="level2"><h2 class="anchored" data-anchor-id="select">2.3 <code>select()</code>
</h2>
<p><strong>Functionality:</strong> Select only the columns (variables) that you want to see. Gets rid of all other columns.</p>
<p>Imagine you are interested to extract the data only for one gene existing in your dataset. You can use the gene name, for example “CD4”, to indicate in select function which column should be selected.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-33_cc45295601b17c34d24d45dc94ebb1c3">
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># choosing the column of CD4 gene</span></span>
<span><span class="va">expMat</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="st">"CD4"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">CD4_column</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In the table we work on, the protein and gene expression information are provided. All the protein markers are labeled with “pAbO” abbreviation, in addition to marker name. We can <strong>select</strong> all the columns that <strong>contain</strong> pAbO in name and save to new table.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-34_e2b6a3a9e93e66b137508390445af04d">
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#choosing all the columns with AbSeq counts and saving to abseq_data</span></span>
<span><span class="va">expMat</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">contains</a></span><span class="op">(</span><span class="st">"pAbO"</span><span class="op">)</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">abseq_data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>All the columns that does not contain the “pAbO” are representing genes. To create a gene data, we can use the contains() function with exclamation mark as in below example.</p>
<p>The “!” is used for negation in R language and is interpreted as “do not”. Below expression will be interpreted as “<strong>select</strong> from expMat columns that <strong>do not</strong> <strong>contain</strong> pAbO.”.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-35_ddab400be74e675ac4331db55cb15968">
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># choosing all columns that does not contain AbSeq counts (mRNA only)</span></span>
<span><span class="va">expMat</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">contains</a></span><span class="op">(</span><span class="st">"pAbO"</span><span class="op">)</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">mrna_data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="mutate" class="level2"><h2 class="anchored" data-anchor-id="mutate">2.4 <code>mutate()</code>
</h2>
<p><strong>Functionality:</strong> Adds new columns or modifies current variables in the dataset.</p>
<p>The mutate function allows in easy way to add new columns in existing tables. Let’s say we want to add column named “measurement_type” and fill with “protein expression” in abseq data table we created in previous step.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-36_8a0675ff8ec85276a922696bc662e7fe">
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#add new column with the same value</span></span>
<span><span class="va">abseq_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>data_modality <span class="op">=</span> <span class="st">"protein_expression"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">abseq_data</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The new column will be added at the end of table. We can show the column by selecting the column name.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-37_739fc9aa8ea2dce166949456003e2c07">
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#display results</span></span>
<span><span class="va">abseq_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="st">"data_modality"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              data_modality
3261781  protein_expression
9155806  protein_expression
4304661  protein_expression
2216902  protein_expression
13160139 protein_expression
11951658 protein_expression</code></pre>
</div>
</div>
<p>Mutate function can be used in combination with other functions.</p>
<p>In this example we will calculate sum of expression count for each cell and store information in table. We summarize the counts from all rows ad add as a new column. The rowSums function is useful for calculating the sum of counts in each row.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-38_3beef63d6d8dabdb19026ffe2b605120">
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">mrna_data</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">mRNA_count_sum</span></span>
<span></span>
<span><span class="co">#display results</span></span>
<span><span class="va">mRNA_count_sum</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> 3261781  9155806  4304661  2216902 13160139 11951658 
   35231    34733    33238    31618    31664    30786 </code></pre>
</div>
</div>
<p>After running rowSums we got for each cell the number of counts. The values can be added in mrna data table as a new column.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-39_9a9b9b46d79650c28b676ed8d0dc8f92">
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mrna_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">mRNA_count_sum</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">mrna_data</span></span>
<span><span class="va">mrna_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         counts
3261781   35231
9155806   34733
4304661   33238
2216902   31618
13160139  31664
11951658  30786</code></pre>
</div>
</div>
<p>The same result can be generated with option below, where the rowSums function is nested inside the mutate function.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-40_0e1d492ec661fed6681383818c523b75">
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mrna_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">mrna_data</span><span class="op">)</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">mrna_data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Notice that mrna_data is given twice in that case as an argument. The mrna_data is taken by mutate function as first argument and then in rowSums function as second argument.</p>
<p>In above example we used pipe to direct the mrna_data to mutate() function as argument. However the mrna_data was not forwarded by pipe to rowSums. For that reason we had to type it in the rowSums function. By default pipe operator forwards data only as first argument to a function. If you want the pipe to forward data as well as second (and next) argument to function you can use a dot.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-41_618893502284a90e2b7f46a38171d548">
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mrna_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">mrna_data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="summarise" class="level2"><h2 class="anchored" data-anchor-id="summarise">2.5 <code>summarise()</code>
</h2>
<p><strong>Functionality:</strong> Collapses all rows and returns a one-row summary.</p>
<p>In the example below, summarize function allow us to calculate the max/min, median and mean values for counts.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-42_1137a06f530011f39c2f324e3d4f4d60">
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mrna_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,    <span class="co">#mean counts</span></span>
<span>                             median <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,    <span class="co">#median counts</span></span>
<span>                             max <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,    <span class="co">#maximum count</span></span>
<span>                             min <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span>   <span class="co">#minimum count</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      mean median    max  min
1 9142.064   4392 105693 1146</code></pre>
</div>
</div>
<p>In above summary we have the overview of the expression range and how minimum and maximum values of molecules expressed differ. Some of cells will have very low transcriptome size in comparison to others.</p>
</section><section id="if_else" class="level2"><h2 class="anchored" data-anchor-id="if_else">2.6 <code>if_else()</code>
</h2>
<p>The if_else() function in R is a conditional function. It allows to test the hypothesis and perform actions based on answer.</p>
<p>Let’s assume we want to label the cells according to transcriptom size. Based on threshold of counts we give a label of “low_transcriptome” or “high_transcriptome” for a cell status. We can make assumption that all cells with sum of counts lower then thousand are low transcriptome cells and remaining cells are categorized as high transcriptome cells.</p>
<p>The if_else function with combination of mutate will help us to achieve the goal.</p>
<p>First we need to know what is the structure of if_else statement. The first argument of if_else function is the hypotheses or the statement. In our case we want to test if count for cell is below or above the threshold. Second argument of a function is an action to perform if the statement is true. Third argument of if_else function is an action to perform when the statement is false.</p>
<p><img src="prerequisite_figures/ifelse.jpg" class="img-fluid"></p>
<p>In our case the hypothesis to test is if counts in a cell are below or above the threshold. We can set the condition to have a counts above 2000. If the counts are above the threshold the statement is true. When the statement is true the action is to set the label as “high_transcriptome”. When the statement is false we set the label as “low_transciptome”.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-43_501e50016dd1db9be7225ffe65c85591">
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">mrna_data</span><span class="op">$</span><span class="va">counts</span> <span class="op">&gt;</span> <span class="fl">2000</span>, <span class="st">"high_transcriptome"</span>, <span class="st">"low_transcriptome"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">cell_status</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In that way we created a cell status for cells based on counts. We should see in a cell_status that was generated two labels. To verify if that is the case we can display unique values of a created vector.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-44_7442f8e732c605cda2e41ac84bb57209">
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">cell_status</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "high_transcriptome" "low_transcriptome" </code></pre>
</div>
</div>
<p>If we would like to add that information as a new column to data table, we can use mutate function as in previous examples.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-45_5d729cdde9098ba644c8100cb9338d83">
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#calcuate and add cell_status to mrna table</span></span>
<span><span class="va">mrna_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>cell_status <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html">if_else</a></span><span class="op">(</span><span class="va">mrna_data</span><span class="op">[</span><span class="st">"counts"</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">2000</span>, <span class="st">"high_transcriptome"</span>, <span class="st">"low_transcriptome"</span><span class="op">)</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">mrna_data</span></span>
<span></span>
<span><span class="co">#display cell status column </span></span>
<span><span class="va">mrna_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">cell_status</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                cell_status
3261781  high_transcriptome
9155806  high_transcriptome
4304661  high_transcriptome
2216902  high_transcriptome
13160139 high_transcriptome
11951658 high_transcriptome</code></pre>
</div>
</div>
</section><section id="count" class="level2"><h2 class="anchored" data-anchor-id="count">2.7 <code>count()</code>
</h2>
<p><strong>Functionality:</strong> Collapses the rows and counts the number of observations per group of values.</p>
<p>The count function is a quick way to have an overview on number of occurrences. For example to we can use <code><a href="https://dplyr.tidyverse.org/reference/count.html">count()</a></code> to see how many cells were categorized as low or high transcriptome cells.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-46_63d0cb4077bd31bca5b03d5c30a6acd3">
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mrna_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html">count</a></span><span class="op">(</span><span class="va">cell_status</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         cell_status    n
1 high_transcriptome 5204
2  low_transcriptome  597</code></pre>
</div>
</div>
</section><section id="group_by" class="level2"><h2 class="anchored" data-anchor-id="group_by">2.8 <code>group_by()</code>
</h2>
<p><strong>Functionality</strong>: Takes existing data and groups specific variables together for future operations.</p>
<p>The function takes an existing table and convert it into grouped table where operations are performed by group.</p>
<p>Let’s have a look into following example and assume that we want to compare the mean and median counts for low and high transcriptome cells. In that case we can group data by cell status and calculate average values for those groups.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-47_e69af62fd32130b873cfaa3493f9a77d">
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mrna_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cell_status</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span>mean <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>            median <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>            n <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>  <span class="co">#calculates the total number of observations</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 4
  cell_status          mean median     n
  &lt;chr&gt;               &lt;dbl&gt;  &lt;dbl&gt; &lt;int&gt;
1 high_transcriptome 10006.  4684.  5204
2 low_transcriptome   1611.  1638    597</code></pre>
</div>
</div>
<p>In the above code, we have grouped by cell status, meaning that calculations performed on our data will account for high transcriptome and low transcriptome cells separately. Following code execution, the console displays the mean count, the median count and the total number of cells (n()) for low transcriptome and for high transcriptome cells (group_by(cell_status)).</p>
</section><section id="filter" class="level2"><h2 class="anchored" data-anchor-id="filter">2.8 <code>filter()</code>
</h2>
<p><strong>Functionality:</strong> Only retain specific rows of data that meet the specified requirement(s).</p>
<p>Filter function display from the data values that meet defined condition. For example, we can filter all cells with low transcriptome cell status and retain only high transcriptome cells.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-48_be2f3066b67041d5e351030c687b89d9">
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mrna_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">cell_status</span> <span class="op">==</span> <span class="st">"high_transcriptome"</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">signal_cells</span></span>
<span></span>
<span><span class="co">#display last column</span></span>
<span><span class="va">signal_cells</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">last_col</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         cell_status
1 high_transcriptome
2 high_transcriptome
3 high_transcriptome
4 high_transcriptome
5 high_transcriptome
6 high_transcriptome</code></pre>
</div>
</div>
</section><section id="rename" class="level2"><h2 class="anchored" data-anchor-id="rename">2.9 <code>rename()</code>
</h2>
<p><strong>Functionality:</strong> Renames a column/variable by name and index.</p>
<p>Renames the cell_status variable as SIGNAL_CELLS (notice that the new label goes first in the argument).</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-49_4b148ada7ed0560c7e85fc8254d40ac5">
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mrna_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>SIGNAL_CELLS <span class="op">=</span> <span class="va">cell_status</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">mrna_data</span></span>
<span></span>
<span><span class="co">#display results</span></span>
<span><span class="va">mrna_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">last_col</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span> <span class="co">#chose last column from mrna_data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               SIGNAL_CELLS
3261781  high_transcriptome
9155806  high_transcriptome
4304661  high_transcriptome
2216902  high_transcriptome
13160139 high_transcriptome
11951658 high_transcriptome</code></pre>
</div>
</div>
</section></section><section id="seurat-object" class="level1"><h1>3. Seurat object</h1>
<p>Seurat object is a data structure designed for storing single-cell genomics data and associated information. We can think about Seurat object as a box with different sections to keep single-cell data in organized manner.</p>
<p>In this chapter we will have a look in what information we can find in Seurat object and how to categorize them.</p>
<section id="seurat-object-structure" class="level2"><h2 class="anchored" data-anchor-id="seurat-object-structure">3.1 Seurat Object Structure</h2>
<p>Seurat object stores information about the single cell multiomics data and the analysis that was done during downstream workflow. As we move through the workflow and make calculations, the results are added to Seurat object and stored in different “boxes”. Those different “boxes” are called slots.</p>
<p>Let’s load the Seurat object into R.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-50_6b62104bc045240d701b9028f33595b1">
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seuratObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"data/precourse_preparation_data/seuratObj.rds"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Loaded Seurat object was created including mRNA counts table obtained from primary analysis. While Seurat object is generated the following information are added to Seurat slots by default: assays, meta data, active assay, active identity, project name and version.</p>
<p><img src="prerequisite_figures/seurat_slots.jpg" class="img-fluid"></p>
<p>The remaining Seurat slots highlighted in grey are empty and filled in with information while it is calculated over analysis.</p>
<p>You can open the Seurat object in separate tab and review its content.</p>
</section><section id="assays" class="level2"><h2 class="anchored" data-anchor-id="assays">3.2 Assays</h2>
<p>The assays slot contain different assays loaded into Seurat object. Those can be for example mRNA expression data or protein expression data. The assay slot is described by additional subslots that stores ‘transformations’ of the data, including raw counts (<code>counts</code> slot), normalized data (<code>data</code> slot), and scaled data for dimensional reduction (<code>scale.data</code> slot).</p>
<p>The assays slots can be accessed by function <code>GetAssayData</code>. In example below we access the <code>counts</code> slot.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-51_9c5e22b80958a52931a9f0ac13cefd6b">
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#Seurat object saved before normalization</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html">GetAssayData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">seuratObj</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="va">.</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span> , <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"6660551"</span>,<span class="st">"2066735"</span>,<span class="st">"9164959"</span>,<span class="st">"3709837"</span><span class="op">)</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>5 x 4 sparse Matrix of class "dgCMatrix"
         6660551 2066735 9164959 3709837
A1BG           .       1       .       .
A1BG.AS1       1       .       1       .
A1CF           .       .       .       .
A2M            .       .       .       2
A2M.AS1        .       .       .       .</code></pre>
</div>
</div>
<p>We can perform normalization and store the normalized counts in <code>data</code> slot.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-52_a582a703504e55937879a99f5513b14a">
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#Seurat object saved after normalization</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span><span class="va">seuratObj</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html">GetAssayData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">.</span>, slot <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="va">.</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span> , <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"6660551"</span>,<span class="st">"2066735"</span>,<span class="st">"9164959"</span>,<span class="st">"3709837"</span><span class="op">)</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>5 x 4 sparse Matrix of class "dgCMatrix"
          6660551   2066735   9164959   3709837
A1BG     .        0.3399948 .         .        
A1BG.AS1 0.330402 .         0.3532289 .        
A1CF     .        .         .         .        
A2M      .        .         .         0.6642082
A2M.AS1  .        .         .         .        </code></pre>
</div>
</div>
<p>By running above example you can see the raw values in data slot are changed due to LogTransform normalization.</p>
<p>Below you can find a useful graphics indicating how data are altered in assays slots in Seurat object after applying different functions related to normalization and scaling the data.</p>
<p><img src="prerequisite_figures/seurat_explained_cropped.png" class="img-fluid"></p>
<p>Credit: Lior Pachter https://twitter.com/lpachter/status/1524413513233575936</p>
</section><section id="meta-data" class="level2"><h2 class="anchored" data-anchor-id="meta-data">3.3 Meta Data</h2>
<p><code>meta.data</code> slot stores information calculated for cells as for example number of features in each cell, number of counts or cell clusters calculated in downstream process.</p>
<p>The meta.data information is stored in a table that can be accessed in Seurat object as below.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-53_c31517240c51bc044edbb7c0fef9f6f4">
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seuratObj</span><span class="op">@</span><span class="va">meta.data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            orig.ident nCount_RNA nFeature_RNA
3261781  SeuratProject      35231         5823
9155806  SeuratProject      34733         5943
4304661  SeuratProject      33238         5480
2216902  SeuratProject      31618         5573
13160139 SeuratProject      31664         5592
11951658 SeuratProject      30786         5457</code></pre>
</div>
</div>
<p>To view list of all available meta.data in Seurat object we can view the column names from the data table.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-54_b51b0031c2cd652b2efd9fc7f5c7795e">
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">seuratObj</span><span class="op">@</span><span class="va">meta.data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "orig.ident"   "nCount_RNA"   "nFeature_RNA"</code></pre>
</div>
</div>
<p>It is possible to add new information to meta data. It can be any information about the cells as experimental batch or sub-population identity.</p>
<div class="cell" data-hash="precourse_preparation_R_cache/html/unnamed-chunk-55_e115fdd34bc5bdd04b44b1aba18d3eaa">
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#add meta data</span></span>
<span><span class="va">seuratObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AddMetaData.html">AddMetaData</a></span><span class="op">(</span><span class="va">seuratObj</span>, <span class="co">#indicate object</span></span>
<span>            <span class="st">"experiment_1"</span>, <span class="co">#metadata to be added</span></span>
<span>            <span class="st">"batch"</span><span class="op">)</span> <span class="co">#column name for metadata</span></span>
<span></span>
<span><span class="co">#display results</span></span>
<span><span class="va">seuratObj</span><span class="op">@</span><span class="va">meta.data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">head</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            orig.ident nCount_RNA nFeature_RNA        batch
3261781  SeuratProject      35231         5823 experiment_1
9155806  SeuratProject      34733         5943 experiment_1
4304661  SeuratProject      33238         5480 experiment_1
2216902  SeuratProject      31618         5573 experiment_1
13160139 SeuratProject      31664         5592 experiment_1
11951658 SeuratProject      30786         5457 experiment_1</code></pre>
</div>
</div>
</section></section><section id="references-and-resources" class="level1"><h1>4. References and Resources</h1>
<p>https://www.geeksforgeeks.org/r-data-frames/</p>
<p>https://bookdown.org/yih_huynh/Guide-to-R-Book/tidyverse.html</p>
<p>https://silviacanelon.com/talk/2020-tour-of-the-tidyverse/</p>
<p>https://oliviergimenez.github.io/intro_tidyverse/</p>
<p>https://jhudatascience.org/tidyversecourse/get-data.html</p>
<p>https://www.analyticsvidhya.com/blog/2019/05/beginner-guide-tidyverse-most-powerful-collection-r-packages-data-science/</p>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>